{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/378530466.py:9: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(ExampleData, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPa</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Positionb</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Effect Allele</th>\n",
       "      <th>EAFc</th>\n",
       "      <th>Overall Breast Cancerd</th>\n",
       "      <th>ER-positivee</th>\n",
       "      <th>ER-negativef</th>\n",
       "      <th>hybrid ER-positiveg</th>\n",
       "      <th>hybrid ER-negativeh</th>\n",
       "      <th>in_23andMe</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_100880328_A_T</td>\n",
       "      <td>1</td>\n",
       "      <td>100880328</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.409678</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_10566215_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>10566215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.329016</td>\n",
       "      <td>-0.0586</td>\n",
       "      <td>-0.0407</td>\n",
       "      <td>-0.1109</td>\n",
       "      <td>-0.0407</td>\n",
       "      <td>-0.1109</td>\n",
       "      <td>True</td>\n",
       "      <td>rs616488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_110198129_CAAA_C</td>\n",
       "      <td>1</td>\n",
       "      <td>110198129</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>0.775478</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>0.0545</td>\n",
       "      <td>0.0266</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_114445880_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>114445880</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.166363</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0642</td>\n",
       "      <td>0.0579</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_118141492_A_C</td>\n",
       "      <td>1</td>\n",
       "      <td>118141492</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.265682</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>0.0417</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>22_39343916_T_A</td>\n",
       "      <td>22</td>\n",
       "      <td>39343916</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.254083</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>0.0326</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>22_40904707_CT_C</td>\n",
       "      <td>22</td>\n",
       "      <td>40904707</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>0.109871</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>0.1160</td>\n",
       "      <td>0.1203</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>22_43433100_C_T</td>\n",
       "      <td>22</td>\n",
       "      <td>43433100</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.114445</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>-0.0585</td>\n",
       "      <td>-0.0515</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>22_45319953_G_A</td>\n",
       "      <td>22</td>\n",
       "      <td>45319953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.416619</td>\n",
       "      <td>-0.0134</td>\n",
       "      <td>-0.0060</td>\n",
       "      <td>-0.0611</td>\n",
       "      <td>-0.0060</td>\n",
       "      <td>-0.0611</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>22_46283297_G_A</td>\n",
       "      <td>22</td>\n",
       "      <td>46283297</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.111687</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>0.0719</td>\n",
       "      <td>0.0993</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>313 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPa  Chromosome  Positionb Reference Allele Effect Allele  \\\n",
       "0       1_100880328_A_T           1  100880328                A             T   \n",
       "1        1_10566215_A_G           1   10566215                A             G   \n",
       "2    1_110198129_CAAA_C           1  110198129             CAAA             C   \n",
       "3       1_114445880_G_A           1  114445880                G             A   \n",
       "4       1_118141492_A_C           1  118141492                A             C   \n",
       "..                  ...         ...        ...              ...           ...   \n",
       "308     22_39343916_T_A          22   39343916                T             A   \n",
       "309    22_40904707_CT_C          22   40904707               CT             C   \n",
       "310     22_43433100_C_T          22   43433100                C             T   \n",
       "311     22_45319953_G_A          22   45319953                G             A   \n",
       "312     22_46283297_G_A          22   46283297                G             A   \n",
       "\n",
       "         EAFc  Overall Breast Cancerd  ER-positivee  ER-negativef  \\\n",
       "0    0.409678                  0.0373        0.0355        0.0160   \n",
       "1    0.329016                 -0.0586       -0.0407       -0.1109   \n",
       "2    0.775478                  0.0458        0.0545        0.0266   \n",
       "3    0.166363                  0.0621        0.0642        0.0579   \n",
       "4    0.265682                  0.0452        0.0417        0.0551   \n",
       "..        ...                     ...           ...           ...   \n",
       "308  0.254083                  0.0407        0.0326        0.0330   \n",
       "309  0.109871                  0.1148        0.1160        0.1203   \n",
       "310  0.114445                 -0.0600       -0.0585       -0.0515   \n",
       "311  0.416619                 -0.0134       -0.0060       -0.0611   \n",
       "312  0.111687                  0.0736        0.0719        0.0993   \n",
       "\n",
       "     hybrid ER-positiveg  hybrid ER-negativeh  in_23andMe      rsid  \n",
       "0                 0.0373               0.0373       False       NaN  \n",
       "1                -0.0407              -0.1109        True  rs616488  \n",
       "2                 0.0458               0.0458       False       NaN  \n",
       "3                 0.0621               0.0621       False       NaN  \n",
       "4                 0.0452               0.0452       False       NaN  \n",
       "..                   ...                  ...         ...       ...  \n",
       "308               0.0407               0.0407       False       NaN  \n",
       "309               0.1148               0.1148       False       NaN  \n",
       "310              -0.0600              -0.0600       False       NaN  \n",
       "311              -0.0060              -0.0611       False       NaN  \n",
       "312               0.0736               0.0736       False       NaN  \n",
       "\n",
       "[313 rows x 13 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the 23andMe raw data file and turn it into a dataframe\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "\n",
    "ExampleData = \"../../Data/23andMe_files/genome_Chad_Wrye_v5_Full_20220921063742.txt\"\n",
    "\n",
    "# Read the text file into a DataFrame\n",
    "ttAndMe = pd.read_csv(ExampleData, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
    "\n",
    "# Find all the positions in ttAndMe that are in PRS313 by identifying columns with the same position and chromosome\n",
    "PRS313 = pd.read_excel('../../Data/PRS313.xlsx')\n",
    "\n",
    "PRS313['in_23andMe'] = PRS313.apply(lambda x: ttAndMe[(ttAndMe['position'] == x['Positionb']) & (ttAndMe['chromosome'] == x['Chromosome'])].shape[0] > 0, axis=1)\n",
    "\n",
    "PRS313[(PRS313['in_23andMe'] == True) & (PRS313[\"Chromosome\"] == 1)]\n",
    "\n",
    "PRS313.to_excel('../../Data/PRS313_with_23andMe.xlsx', index=False)\n",
    "\n",
    "PRS313"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing file 12330.23andme.10166.txt: 'utf-8' codec can't decode byte 0xff in position 15: invalid start byte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing file 12335.23andme.10172.txt: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing file 12373.23andme.10203.txt: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_57645/679208585.py:18: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 28\u001b[0m\n\u001b[1;32m     26\u001b[0m     chromosome \u001b[38;5;241m=\u001b[39m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChromosome\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     27\u001b[0m     position \u001b[38;5;241m=\u001b[39m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPositionb\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m---> 28\u001b[0m     snp_presence[snp_location] \u001b[38;5;241m=\u001b[39m ttAndMe[(ttAndMe[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mposition\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m position) \u001b[38;5;241m&\u001b[39m (\u001b[43mttAndMe\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mchromosome\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mchromosome\u001b[49m)]\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m# Count the total number of SNPs present\u001b[39;00m\n\u001b[1;32m     31\u001b[0m total_snps_present \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m(snp_presence\u001b[38;5;241m.\u001b[39mvalues())\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/ops/common.py:76\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[1;32m     74\u001b[0m other \u001b[38;5;241m=\u001b[39m item_from_zerodim(other)\n\u001b[0;32m---> 76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mother\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/arraylike.py:40\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__eq__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m---> 40\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cmp_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meq\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/series.py:5799\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   5796\u001b[0m lvalues \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values\n\u001b[1;32m   5797\u001b[0m rvalues \u001b[38;5;241m=\u001b[39m extract_array(other, extract_numpy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, extract_range\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m-> 5799\u001b[0m res_values \u001b[38;5;241m=\u001b[39m \u001b[43mops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcomparison_op\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5801\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_result(res_values, name\u001b[38;5;241m=\u001b[39mres_name)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/ops/array_ops.py:346\u001b[0m, in \u001b[0;36mcomparison_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    343\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[1;32m    345\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m lvalues\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(rvalues, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 346\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mcomp_method_OBJECT_ARRAY\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    348\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    349\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/ops/array_ops.py:131\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[0;34m(op, x, y)\u001b[0m\n\u001b[1;32m    129\u001b[0m     result \u001b[38;5;241m=\u001b[39m libops\u001b[38;5;241m.\u001b[39mvec_compare(x\u001b[38;5;241m.\u001b[39mravel(), y\u001b[38;5;241m.\u001b[39mravel(), op)\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 131\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mlibops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscalar_compare\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mravel\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\u001b[38;5;241m.\u001b[39mreshape(x\u001b[38;5;241m.\u001b[39mshape)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Set the path to the folder containing the 23andMe files\n",
    "folder_path = \"../../Data/23andMe_files\"\n",
    "\n",
    "# Read the PRS313 file\n",
    "PRS313 = pd.read_excel('../../Data/PRS313.xlsx')\n",
    "\n",
    "# Initialize a DataFrame to store the results\n",
    "result_df = pd.DataFrame(columns=PRS313['SNPa'].tolist() + ['total_snps_present', 'total_positions'])\n",
    "\n",
    "# Initialize a DataFrame to store the unique positions from all 23andMe files\n",
    "unique_positions_df = pd.DataFrame(columns=['rsid', 'chromosome', 'position'])\n",
    "\n",
    "# Initialize a list to store the positions from each 23andMe file\n",
    "positions_list = []\n",
    "\n",
    "# Iterate over each file in the folder\n",
    "for filename in os.listdir(folder_path):\n",
    "    file_path = os.path.join(folder_path, filename)\n",
    "    try:\n",
    "        # Read the 23andMe raw data file into a DataFrame\n",
    "        ttAndMe = pd.read_csv(file_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
    "\n",
    "        # If the number of positions is greater than 700000 or less than 600000, delete the file and move on\n",
    "        if ttAndMe.shape[0] > 700000 or ttAndMe.shape[0] < 600000:\n",
    "            os.remove(file_path)\n",
    "            continue\n",
    "\n",
    "        # Initialize a dictionary to store the presence of each SNP\n",
    "        snp_presence = {}\n",
    "\n",
    "        # Check the presence of each SNP in the 23andMe file\n",
    "        for _, row in PRS313.iterrows():\n",
    "            snp_location = row['SNPa']\n",
    "            chromosome = row['Chromosome']\n",
    "            position = row['Positionb']\n",
    "            snp_presence[snp_location] = ttAndMe[(ttAndMe['position'] == position) & (ttAndMe['chromosome'] == chromosome)].shape[0] > 0\n",
    "\n",
    "        # Count the total number of SNPs present\n",
    "        total_snps_present = sum(snp_presence.values())\n",
    "\n",
    "        # Count the total number of positions in the 23andMe file\n",
    "        total_positions = ttAndMe.shape[0]\n",
    "\n",
    "        # Add the results to the DataFrame\n",
    "        result_df.loc[filename] = [snp_presence[snp_location] for snp_location in PRS313['SNPa'].tolist()] + [total_snps_present, total_positions]\n",
    "\n",
    "        # Append the positions from the current 23andMe file to the positions_list\n",
    "        positions_list.append(ttAndMe[['rsid', 'chromosome', 'position']])\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing file {filename}: {e}\")\n",
    "        continue\n",
    "\n",
    "# Concatenate all the positions from each 23andMe file\n",
    "all_positions_df = pd.concat(positions_list, ignore_index=True)\n",
    "\n",
    "# Get the unique positions from all 23andMe files\n",
    "unique_positions_df = all_positions_df.drop_duplicates(subset=['chromosome', 'position', 'rsid'])\n",
    "\n",
    "# Find the intersection of positions present in all 23andMe files\n",
    "intersection_positions_df = all_positions_df.groupby(['chromosome', 'position']).filter(lambda x: len(x) == len(positions_list))\n",
    "\n",
    "# Save the results to Excel files\n",
    "result_df.to_excel('../../Data/PRS313_23andMe_results.xlsx', index_label='filename')\n",
    "unique_positions_df.to_excel('../../Data/23andMe_unique_positions.xlsx', index=False)\n",
    "intersection_positions_df.to_excel('../../Data/23andMe_intersection_positions.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frequency Table for Total SNPs Present:\n",
      "total_snps_present_bins\n",
      "0-59        9\n",
      "60-70      28\n",
      "71-150      3\n",
      "151-200     0\n",
      "201-250     0\n",
      "251-300     1\n",
      "301-350     0\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Frequency Table for Total Positions:\n",
      "total_positions_bins\n",
      "0-550k        1\n",
      "550k-700k    34\n",
      "700k-1.5M     4\n",
      "1.5M-2M       0\n",
      "2M-2.5M       1\n",
      "2.5M-10M      1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Create frequency table with bins for total_snps_present\n",
    "bins = [0, 59, 71, 150, 200, 250, 300, 350]\n",
    "labels = ['0-59', '60-70', '71-150', '151-200', '201-250', '251-300', '301-350']\n",
    "result_df['total_snps_present_bins'] = pd.cut(result_df['total_snps_present'], bins=bins, labels=labels, right=False)\n",
    "total_snps_present_freq = result_df['total_snps_present_bins'].value_counts().sort_index()\n",
    "\n",
    "print(\"Frequency Table for Total SNPs Present:\")\n",
    "print(total_snps_present_freq)\n",
    "print()\n",
    "\n",
    "# Create frequency table with bins for total_positions\n",
    "bins = [0, 550000, 700000, 1500000, 2000000, 2500000, 10000000]\n",
    "labels = ['0-550k', '550k-700k', '700k-1.5M', '1.5M-2M', '2M-2.5M', '2.5M-10M']\n",
    "result_df['total_positions_bins'] = pd.cut(result_df['total_positions'], bins=bins, labels=labels, right=False)\n",
    "total_positions_freq = result_df['total_positions_bins'].value_counts().sort_index()\n",
    "\n",
    "print(\"Frequency Table for Total Positions:\")\n",
    "print(total_positions_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "68.29268292682927"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the percentage of present snps between 60-70\n",
    "total_snps_present_freq['60-70'] / result_df.shape[0] * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert all the 23andMe files to text files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renamed 12367.23andme.10198 to 12367.txt\n",
      "Renamed 12312.23andme.10143 to 12312.txt\n",
      "Renamed 12343.23andme.10178 to 12343.txt\n",
      "Renamed 12232.23andme.10180 to 12232.txt\n",
      "Renamed 12325.23andme.10160 to 12325.txt\n",
      "Renamed 12332.23andme.10169 to 12332.txt\n",
      "Renamed 12344.23andme.10179 to 12344.txt\n",
      "Renamed 12334.23andme.10171 to 12334.txt\n",
      "Renamed 3033.23andme.10150 to 3033.txt\n",
      "Renamed 12335.23andme.10172 to 12335.txt\n",
      "Renamed 12356.23andme.10187 to 12356.txt\n",
      "Renamed 12370.23andme.10200 to 12370.txt\n",
      "Renamed 12320.23andme.10152 to 12320.txt\n",
      "Renamed 12368.23andme.10199 to 12368.txt\n",
      "Renamed 12348.23andme.10183 to 12348.txt\n",
      "Renamed 12337.23andme.10181 to 12337.txt\n",
      "Renamed 12363.23andme.10193 to 12363.txt\n",
      "Renamed 4015.23andme.10141 to 4015.txt\n",
      "Renamed 12371.23andme.10201 to 12371.txt\n",
      "Renamed 12309.23andme.10142 to 12309.txt\n",
      "Renamed 12308.23andme.10140 to 12308.txt\n",
      "Renamed 12365.23andme.10196 to 12365.txt\n",
      "Renamed 12330.23andme.10166 to 12330.txt\n",
      "Renamed 12342.23andme.10176 to 12342.txt\n",
      "Renamed 12351.23andme.10185 to 12351.txt\n",
      "Renamed 12319.23andme.10151 to 12319.txt\n",
      "Renamed 12373.23andme.10203 to 12373.txt\n",
      "Renamed 12339.23andme.10173 to 12339.txt\n"
     ]
    }
   ],
   "source": [
    "# Set the path to the folder containing the 23andMe files\n",
    "folder_path = \"../../Data/23andMe_files\"\n",
    "\n",
    "# Convert all the 23andMe files to .txt format if they are not already in that format\n",
    "\n",
    "for filename in os.listdir(folder_path):\n",
    "    file_path = os.path.join(folder_path, filename)\n",
    "    \n",
    "    # Check if the file is already in .txt format\n",
    "    if not filename.endswith('.txt'):\n",
    "        new_file_path = os.path.join(folder_path, filename + '.txt')\n",
    "        os.rename(file_path, new_file_path)\n",
    "        print(f\"Renamed {filename} to {filename.split('.')[0] + '.txt'}\")\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
