{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "def read_vcf_header(file_path):\n",
    "    \"\"\"Reads the VCF file header (column names) and returns it.\"\"\"\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('#CHROM'):\n",
    "                return line.strip().split('\\t')\n",
    "    return []\n",
    "\n",
    "def extract_snp_data(file_path, header):\n",
    "    \"\"\"Extracts SNP data from a VCF file.\"\"\"\n",
    "    snp_data = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('#'):\n",
    "                fields = line.strip().split('\\t')\n",
    "                chrom, pos, ref, alt = fields[0], fields[1], fields[3], fields[4]\n",
    "                snp_id = f\"chr{chrom}_{pos}_{ref}_{alt}\"\n",
    "                genotype_data = fields[9:]  # Genotype data starts from the 10th column\n",
    "                snp_data.append((snp_id, genotype_data))\n",
    "    return snp_data\n",
    "\n",
    "def process_vcf_files(directory):\n",
    "    \"\"\"Processes all VCF files in the specified directory.\"\"\"\n",
    "    vcf_files = glob.glob(os.path.join(directory, '*.vcf'))\n",
    "    all_snp_data = []\n",
    "    patient_ids = None\n",
    "    \n",
    "    for file_path in vcf_files:\n",
    "        if patient_ids is None:\n",
    "            patient_ids = read_vcf_header(file_path)[9:]  # Assuming the first 9 columns are standard VCF columns\n",
    "        snp_data = extract_snp_data(file_path, patient_ids)\n",
    "        all_snp_data.extend(snp_data)\n",
    "\n",
    "    # Convert the collected SNP data to a DataFrame\n",
    "    data_dict = {snp_id: genotypes for snp_id, genotypes in all_snp_data}\n",
    "    df = pd.DataFrame(data_dict, index=patient_ids).transpose()\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: float64)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bcftools.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace 'your_directory_path' with the path to your directory containing the VCF files\n",
    "directory_path = '../../Data/bcftools_results/'\n",
    "df_bcftools = process_vcf_files(directory_path).transpose()\n",
    "\n",
    "\n",
    "# If you want to save this DataFrame to a CSV file for further analysis\n",
    "df_bcftools.to_csv('concatenated_snps.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of VCF files: 0\n"
     ]
    }
   ],
   "source": [
    "# List number of files in the directory\n",
    "vcf_files = glob.glob(os.path.join(directory_path, '*.vcf'))\n",
    "print(f\"Number of VCF files: {len(vcf_files)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['chr22_19707371', 'chr22_19766137', 'chr22_29114293', 'chr22_29121087',\n",
      "       'chr22_29135543', 'chr22_29203724', 'chr22_29551872', 'chr22_38583315',\n",
      "       'chr22_39343916', 'chr22_40904707', 'chr22_43433100', 'chr22_45319942',\n",
      "       'chr22_45319953', 'chr22_46283297'],\n",
      "      dtype='object')\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "# Get all chromosome 10 SNPs\n",
    "chrom_10_snps = df_bcftools.columns[df_bcftools.columns.str.startswith('chr22_')]\n",
    "print(chrom_10_snps)\n",
    "print(len(chrom_10_snps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Replace 'your_directory_path' with the path to your directory containing the TSV files\n",
    "directory_path = '../Grep_tsv_files/'\n",
    "\n",
    "import os\n",
    "\n",
    "def extract_alleles(directory):\n",
    "    \n",
    "    files = {}\n",
    "    for filename in os.listdir(directory):\n",
    "        alleles = []\n",
    "\n",
    "        if filename.endswith(\".vcf\"):\n",
    "            filepath = os.path.join(directory, filename)\n",
    "            with open(filepath, \"r\") as file:\n",
    "                line = file.readline().strip()\n",
    "                fields = line.split(\"\\t\")\n",
    "                \n",
    "                if len(fields) >= 10:\n",
    "                    genotypes = fields[9:]\n",
    "                    for genotype in genotypes:\n",
    "                        \n",
    "                        alleles.append(genotype.replace(\"|\", \"\"))\n",
    "            files[\"chr\" + filename.split(\".\")[0]] = alleles\n",
    "    \n",
    "    return pd.DataFrame(files).T\n",
    "\n",
    "df_grep = extract_alleles(directory_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'chr22_40904707'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexes/base.py:3790\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3789\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3790\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3791\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:152\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:181\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chr22_40904707'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m snp_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr22_40904707\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Get the genotypes for the selected SNP from both dataframes\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m genotypes_bcftools \u001b[38;5;241m=\u001b[39m \u001b[43mdf_bcftools\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43msnp_id\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m      6\u001b[0m genotypes_plink \u001b[38;5;241m=\u001b[39m df_plink\u001b[38;5;241m.\u001b[39mloc[snp_id]\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m      7\u001b[0m genotypes_grep \u001b[38;5;241m=\u001b[39m df_grep\u001b[38;5;241m.\u001b[39mloc[snp_id]\u001b[38;5;241m.\u001b[39mvalues\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexing.py:1153\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1150\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1152\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m-> 1153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexing.py:1393\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1391\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1392\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1393\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexing.py:1343\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1341\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: AxisInt):\n\u001b[1;32m   1342\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1343\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/generic.py:4236\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4234\u001b[0m             new_index \u001b[38;5;241m=\u001b[39m index[loc]\n\u001b[1;32m   4235\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4236\u001b[0m     loc \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4238\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(loc, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   4239\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m loc\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mbool_:\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexes/base.py:3797\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3792\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3793\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3794\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3795\u001b[0m     ):\n\u001b[1;32m   3796\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3797\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3798\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3799\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3801\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chr22_40904707'"
     ]
    }
   ],
   "source": [
    "\n",
    "# Select a SNP ID that is present in both dataframes\n",
    "snp_id = 'chr22_40904707'\n",
    "\n",
    "# Get the genotypes for the selected SNP from both dataframes\n",
    "genotypes_bcftools = df_bcftools.loc[snp_id].values\n",
    "genotypes_plink = df_plink.loc[snp_id].values\n",
    "genotypes_grep = df_grep.loc[snp_id].values\n",
    "\n",
    "\n",
    "# Replace the '/' and '|' characters in the genotypes\n",
    "genotypes_bcftools = [genotype.replace('/', '').replace('|', '') for genotype in genotypes_bcftools]\n",
    "genotypes_plink = [genotype.replace('/', '').replace('|', '') for genotype in genotypes_plink]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The genotypes for SNP chr22_40904707 are not equal in both dataframes.\n",
      "Patient HG00119: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00132: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00136: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00158: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00171: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00186: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00251: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00255: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00267: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00271: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00274: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00280: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00281: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00304: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00306: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00309: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00313: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00326: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00327: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00341: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00344: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00357: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00360: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00364: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00367: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00379: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00380: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00382: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00384: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00403: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00410: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00421: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00458: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00476: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00478: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00531: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00556: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00557: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00583: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00599: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00613: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00625: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00628: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00634: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00650: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00653: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00654: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00663: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00683: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00684: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00690: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00693: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00699: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG00844: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01048: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01049: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01122: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01134: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01136: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01142: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01187: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01251: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01348: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01356: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01438: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01440: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01455: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01500: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01503: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01605: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01617: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01624: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01630: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01697: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01707: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01768: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01796: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01801: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01810: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01812: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01841: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01849: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01878: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01880: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01882: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01926: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01933: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG01950: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02010: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02019: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02023: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02032: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02035: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02061: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02064: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02067: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02073: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02082: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02086: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02087: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02088: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02107: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02108: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02131: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02144: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02151: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02180: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02238: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02275: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02278: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02304: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02312: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02314: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02323: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02325: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02332: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02348: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02391: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02420: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02445: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02450: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02461: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02477: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02491: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02494: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02574: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02582: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02646: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02702: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02703: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02721: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02775: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02784: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02790: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02808: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02836: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02860: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02878: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02923: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02977: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG02981: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03048: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03063: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03084: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03085: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03108: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03111: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03237: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03267: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03367: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03385: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03437: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03461: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03491: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03575: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03595: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03598: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03603: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03625: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03629: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03644: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03673: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03702: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03706: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03709: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03730: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03767: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03773: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03781: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03897: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03900: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03902: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03922: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03945: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03951: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03989: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG03991: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04014: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04018: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04033: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04039: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04171: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04186: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04194: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04195: BCFTOOLS - 10, PLINK - 01\n",
      "Patient HG04235: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA06984: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA11992: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12154: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12249: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12348: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12413: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12750: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12762: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12878: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA12890: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18486: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18498: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18504: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18511: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18526: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18542: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18544: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18546: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18563: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18567: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18579: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18593: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18599: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18605: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18606: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18614: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18615: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18617: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18619: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18629: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18632: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18633: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18637: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18639: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18640: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18642: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18757: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18870: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18945: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18949: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18950: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18951: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18961: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18964: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18966: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18971: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18972: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18973: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18974: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18979: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA18984: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19000: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19003: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19062: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19063: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19066: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19068: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19072: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19080: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19090: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19092: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19095: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19098: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19102: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19113: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19121: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19175: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19190: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19200: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19206: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19257: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19307: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19312: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19372: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19375: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19376: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19378: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19397: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19399: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19428: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19455: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19720: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19725: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19740: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA19920: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20127: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20278: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20281: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20298: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20314: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20321: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20340: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20362: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20512: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20535: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20539: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20586: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20799: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20806: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20811: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20812: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20875: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20892: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA20899: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21088: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21089: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21090: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21091: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21105: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21118: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21124: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21127: BCFTOOLS - 10, PLINK - 01\n",
      "Patient NA21137: BCFTOOLS - 10, PLINK - 01\n",
      "Total number of patients with different genotypes: 299\n"
     ]
    }
   ],
   "source": [
    "# Check if the genotypes are equal\n",
    "genotypes_equal = all(g1 == g2 for g1, g2 in zip(genotypes_bcftools, genotypes_plink))\n",
    "\n",
    "if genotypes_equal:\n",
    "    print(f\"The genotypes for SNP {snp_id} are equal in both dataframes.\")\n",
    "else:\n",
    "    print(f\"The genotypes for SNP {snp_id} are not equal in both dataframes.\")\n",
    "    # Get the particular patients that are not equal\n",
    "    counter = 0\n",
    "    for patient_id, g1, g2 in zip(df_bcftools.columns, genotypes_bcftools, genotypes_plink):\n",
    "        if g1 != g2:\n",
    "            print(f\"Patient {patient_id}: BCFTOOLS - {g1}, PLINK - {g2}\")\n",
    "            counter += 1\n",
    "    print(f\"Total number of patients with different genotypes: {counter}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The genotypes for SNP chr22_40904707 are equal in both dataframes.\n"
     ]
    }
   ],
   "source": [
    "# Check if the genotypes are equal between bcftools and grep\n",
    "\n",
    "genotypes_equal = all(g1 == g2 for g1, g2 in zip(genotypes_bcftools, genotypes_grep))\n",
    "\n",
    "if genotypes_equal:\n",
    "    \n",
    "    print(f\"The genotypes for SNP {snp_id} are equal in both dataframes.\")\n",
    "else:\n",
    "    print(f\"The genotypes for SNP {snp_id} are not equal in both dataframes.\")\n",
    "    # Get the particular patients that are not equal\n",
    "    counter = 0\n",
    "    for patient_id, g1, g2 in zip(df_bcftools.columns, genotypes_bcftools, genotypes_grep):\n",
    "        if g1 != g2:\n",
    "            print(f\"Patient {patient_id}: BCFTOOLS - {g1}, GREP - {g2}\")\n",
    "            counter += 1\n",
    "    print(f\"Total number of patients with different genotypes: {counter}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 16\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# Apply the function to the DataFrame index and create a list\u001b[39;00m\n\u001b[0;32m---> 16\u001b[0m formatted_list \u001b[38;5;241m=\u001b[39m [extract_format(index) \u001b[38;5;28;01mfor\u001b[39;00m index \u001b[38;5;129;01min\u001b[39;00m \u001b[43mdf\u001b[49m\u001b[38;5;241m.\u001b[39mindex]\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28mprint\u001b[39m(formatted_list)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "import re\n",
    "# Regular expression to extract chromosome number and position\n",
    "pattern = r\"chr(\\d+)_([0-9]+)\"\n",
    "\n",
    "# Function to apply the pattern and format the string\n",
    "def extract_format(s):\n",
    "    match = re.match(pattern, s)\n",
    "    if match:\n",
    "        chromosome = match.group(1)\n",
    "        position = match.group(2)\n",
    "        return f\"{chromosome}_{position}\"\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Apply the function to the DataFrame index and create a list\n",
    "formatted_list = [extract_format(index) for index in df.index]\n",
    "\n",
    "print(formatted_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "406"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(formatted_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the specific column from the Excel file\n",
    "PRS_313_mutations = pd.read_excel(\"../PRS313.xlsx\", usecols=[\"SNPa\"])\n",
    "\n",
    "# Function to remove everything after the second underscore\n",
    "def remove_after_second_underscore(s):\n",
    "    parts = s.split('_')\n",
    "    if len(parts) > 2:\n",
    "        return '_'.join(parts[:2])\n",
    "    else:\n",
    "        return s\n",
    "\n",
    "# Apply the function to each string in the column\n",
    "PRS_313_mutations['SNPa'] = PRS_313_mutations['SNPa'].apply(remove_after_second_underscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      1_100880328\n",
       "1       1_10566215\n",
       "2      1_110198129\n",
       "3      1_114445880\n",
       "4      1_118141492\n",
       "          ...     \n",
       "308    22_39343916\n",
       "309    22_40904707\n",
       "310    22_43433100\n",
       "311    22_45319953\n",
       "312    22_46283297\n",
       "Name: SNPa, Length: 313, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PRS_313_mutations.SNPa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def difference_list(list_a, list_b):\n",
    "    \"\"\"\n",
    "    Returns elements that are in list_a but not in list_b.\n",
    "    \n",
    "    Parameters:\n",
    "    - list_a: First list.\n",
    "    - list_b: Second list to compare against.\n",
    "    \n",
    "    Returns:\n",
    "    A new list with elements that are in list_a but not in list_b.\n",
    "    \"\"\"\n",
    "    return [item for item in list_a if item not in list_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_SNPs = difference_list(PRS_313_mutations.SNPa,formatted_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../positions/missing_positions_2.txt', 'w') as file:\n",
    "    for item in missing_SNPs:\n",
    "        chromosome, position = item.split('_')\n",
    "        file.write(f\"{chromosome} {position}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
