{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use LD Proxy API to find the relevant positions for each of the PRS313 SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPa</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Positionb</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Effect Allele</th>\n",
       "      <th>EAFc</th>\n",
       "      <th>Overall Breast Cancerd</th>\n",
       "      <th>ER-positivee</th>\n",
       "      <th>ER-negativef</th>\n",
       "      <th>hybrid ER-positiveg</th>\n",
       "      <th>hybrid ER-negativeh</th>\n",
       "      <th>in_23andMe</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_100880328_A_T</td>\n",
       "      <td>1</td>\n",
       "      <td>100880328</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.409678</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1_100880328_A_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_10566215_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>10566215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.329016</td>\n",
       "      <td>-0.0586</td>\n",
       "      <td>-0.0407</td>\n",
       "      <td>-0.1109</td>\n",
       "      <td>-0.0407</td>\n",
       "      <td>-0.1109</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1_10566215_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_110198129_CAAA_C</td>\n",
       "      <td>1</td>\n",
       "      <td>110198129</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>0.775478</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>0.0545</td>\n",
       "      <td>0.0266</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>0.0458</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1_110198129_CAAA_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_114445880_G_A</td>\n",
       "      <td>1</td>\n",
       "      <td>114445880</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.166363</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0642</td>\n",
       "      <td>0.0579</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>0.0621</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1_114445880_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_118141492_A_C</td>\n",
       "      <td>1</td>\n",
       "      <td>118141492</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.265682</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>0.0417</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>0.0452</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1_118141492_A_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>22_39343916_T_A</td>\n",
       "      <td>22</td>\n",
       "      <td>39343916</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.254083</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>0.0326</td>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22_39343916_T_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>22_40904707_CT_C</td>\n",
       "      <td>22</td>\n",
       "      <td>40904707</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>0.109871</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>0.1160</td>\n",
       "      <td>0.1203</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>0.1148</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22_40904707_CT_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>22_43433100_C_T</td>\n",
       "      <td>22</td>\n",
       "      <td>43433100</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.114445</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>-0.0585</td>\n",
       "      <td>-0.0515</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>-0.0600</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22_43433100_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>22_45319953_G_A</td>\n",
       "      <td>22</td>\n",
       "      <td>45319953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.416619</td>\n",
       "      <td>-0.0134</td>\n",
       "      <td>-0.0060</td>\n",
       "      <td>-0.0611</td>\n",
       "      <td>-0.0060</td>\n",
       "      <td>-0.0611</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22_45319953_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>22_46283297_G_A</td>\n",
       "      <td>22</td>\n",
       "      <td>46283297</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.111687</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>0.0719</td>\n",
       "      <td>0.0993</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>0.0736</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22_46283297_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>313 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SNPa  Chromosome  Positionb Reference Allele Effect Allele  \\\n",
       "0       1_100880328_A_T           1  100880328                A             T   \n",
       "1        1_10566215_A_G           1   10566215                A             G   \n",
       "2    1_110198129_CAAA_C           1  110198129             CAAA             C   \n",
       "3       1_114445880_G_A           1  114445880                G             A   \n",
       "4       1_118141492_A_C           1  118141492                A             C   \n",
       "..                  ...         ...        ...              ...           ...   \n",
       "308     22_39343916_T_A          22   39343916                T             A   \n",
       "309    22_40904707_CT_C          22   40904707               CT             C   \n",
       "310     22_43433100_C_T          22   43433100                C             T   \n",
       "311     22_45319953_G_A          22   45319953                G             A   \n",
       "312     22_46283297_G_A          22   46283297                G             A   \n",
       "\n",
       "         EAFc  Overall Breast Cancerd  ER-positivee  ER-negativef  \\\n",
       "0    0.409678                  0.0373        0.0355        0.0160   \n",
       "1    0.329016                 -0.0586       -0.0407       -0.1109   \n",
       "2    0.775478                  0.0458        0.0545        0.0266   \n",
       "3    0.166363                  0.0621        0.0642        0.0579   \n",
       "4    0.265682                  0.0452        0.0417        0.0551   \n",
       "..        ...                     ...           ...           ...   \n",
       "308  0.254083                  0.0407        0.0326        0.0330   \n",
       "309  0.109871                  0.1148        0.1160        0.1203   \n",
       "310  0.114445                 -0.0600       -0.0585       -0.0515   \n",
       "311  0.416619                 -0.0134       -0.0060       -0.0611   \n",
       "312  0.111687                  0.0736        0.0719        0.0993   \n",
       "\n",
       "     hybrid ER-positiveg  hybrid ER-negativeh  in_23andMe  \\\n",
       "0                 0.0373               0.0373       False   \n",
       "1                -0.0407              -0.1109        True   \n",
       "2                 0.0458               0.0458       False   \n",
       "3                 0.0621               0.0621       False   \n",
       "4                 0.0452               0.0452       False   \n",
       "..                   ...                  ...         ...   \n",
       "308               0.0407               0.0407       False   \n",
       "309               0.1148               0.1148       False   \n",
       "310              -0.0600              -0.0600       False   \n",
       "311              -0.0060              -0.0611       False   \n",
       "312               0.0736               0.0736       False   \n",
       "\n",
       "                       pos  \n",
       "0       chr1_100880328_A_T  \n",
       "1        chr1_10566215_A_G  \n",
       "2    chr1_110198129_CAAA_C  \n",
       "3       chr1_114445880_G_A  \n",
       "4       chr1_118141492_A_C  \n",
       "..                     ...  \n",
       "308     chr22_39343916_T_A  \n",
       "309    chr22_40904707_CT_C  \n",
       "310     chr22_43433100_C_T  \n",
       "311     chr22_45319953_G_A  \n",
       "312     chr22_46283297_G_A  \n",
       "\n",
       "[313 rows x 13 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import os\n",
    "import io\n",
    "\n",
    "\n",
    "# API key\n",
    "api_key = \"ac16be4ad92d\"\n",
    "\n",
    "# Base URL for LDProxy API\n",
    "base_url = \"https://ldlink.nih.gov/LDlinkRest/ldproxy\"\n",
    "\n",
    "population = \"ALL\"\n",
    "\n",
    "window = 1000000\n",
    "\n",
    "r2_threshold = 0.01\n",
    "\n",
    "# Read the chromosome positions from the text file\n",
    "PRS313_LD = pd.read_excel(\"../../Data/PRS313_with_23andMe.xlsx\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create a folder to store the CSV files\n",
    "output_folder = \"../ld_variants\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Iterate over each position\n",
    "for sample in PRS313_LD:\n",
    "    if sample[\"in_23andMe\"] == True:\n",
    "         continue\n",
    "\n",
    "    # Extract chromosome and position from the line\n",
    "    chrom = sample[\"Chromosome\"]\n",
    "    pos = sample[\"Positionb\"]\n",
    "    \n",
    "    # Construct the API request URL\n",
    "    url = f\"{base_url}?var={chrom}:{pos}&pop={population}&r2_d=r2&window={window}&genome_build=grch37&token={api_key}\"\n",
    "    \n",
    "    # Send the API request\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "            # Create a StringIO object from the data\n",
    "            data_io = io.StringIO(response.text)\n",
    "\n",
    "            # Read the data into a DataFrame using read_csv\n",
    "            df = pd.read_csv(data_io, sep='\\\\t')\n",
    "            \n",
    "            # # Filter variants with high LD scores (e.g., R2 >= 0.8)\n",
    "            # high_ld_variants = df[df[\"R2\"].astype(float) >= 0.8]\n",
    "            \n",
    "            # Generate a unique filename for the CSV file\n",
    "            output_file = os.path.join(output_folder, f\"{chrom}_{pos}.csv\")\n",
    "            \n",
    "            # Save the high LD variants to a CSV file\n",
    "            df.to_csv(output_file, index=False)\n",
    "            \n",
    "            print(f\"Saved high LD variants for {chrom}:{pos} to {output_file}\")\n",
    "    else:\n",
    "        print(f\"Failed to retrieve data for {chrom}:{pos}. Status code: {response.status_code}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom, pos = positions[0].strip().split()\n",
    "\n",
    "# Construct the API request URL\n",
    "url = f\"{base_url}?var={chrom}:{pos}&pop={population}&r2_d=r2&window={window}&genome_build=grch37&token={api_key}\"\n",
    "\n",
    "# Send the API request\n",
    "response = requests.get(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_16385/1453726256.py:12: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  df = pd.read_csv(data_io, sep='\\\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RS_Number</th>\n",
       "      <th>Coord</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Distance</th>\n",
       "      <th>Dprime</th>\n",
       "      <th>R2</th>\n",
       "      <th>Correlated_Alleles</th>\n",
       "      <th>FORGEdb</th>\n",
       "      <th>RegulomeDB</th>\n",
       "      <th>Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs612683</td>\n",
       "      <td>chr1:100880328</td>\n",
       "      <td>(A/T)</td>\n",
       "      <td>0.4004</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>A=A,T=T</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2b</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12568038</td>\n",
       "      <td>chr1:100879914</td>\n",
       "      <td>(C/T)</td>\n",
       "      <td>0.3415</td>\n",
       "      <td>-414</td>\n",
       "      <td>0.8908</td>\n",
       "      <td>0.6162</td>\n",
       "      <td>A=C,T=T</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs12564838</td>\n",
       "      <td>chr1:100883188</td>\n",
       "      <td>(A/G)</td>\n",
       "      <td>0.3496</td>\n",
       "      <td>2860</td>\n",
       "      <td>0.8609</td>\n",
       "      <td>0.5968</td>\n",
       "      <td>A=A,T=G</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs7517704</td>\n",
       "      <td>chr1:100885419</td>\n",
       "      <td>(A/G)</td>\n",
       "      <td>0.3494</td>\n",
       "      <td>5091</td>\n",
       "      <td>0.8609</td>\n",
       "      <td>0.5962</td>\n",
       "      <td>A=A,T=G</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs12760924</td>\n",
       "      <td>chr1:100898600</td>\n",
       "      <td>(A/T)</td>\n",
       "      <td>0.3546</td>\n",
       "      <td>18272</td>\n",
       "      <td>0.8357</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>A=A,T=T</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1214</th>\n",
       "      <td>rs111334083</td>\n",
       "      <td>chr1:100617033</td>\n",
       "      <td>(G/A)</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>-263295</td>\n",
       "      <td>0.8351</td>\n",
       "      <td>0.0101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>rs141366089</td>\n",
       "      <td>chr1:100950426</td>\n",
       "      <td>(-/T)</td>\n",
       "      <td>0.0575</td>\n",
       "      <td>70098</td>\n",
       "      <td>0.4970</td>\n",
       "      <td>0.0101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216</th>\n",
       "      <td>rs10493930</td>\n",
       "      <td>chr1:101079706</td>\n",
       "      <td>(T/C)</td>\n",
       "      <td>0.0833</td>\n",
       "      <td>199378</td>\n",
       "      <td>0.4070</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>rs57411954</td>\n",
       "      <td>chr1:100712556</td>\n",
       "      <td>(AC/-)</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>-167772</td>\n",
       "      <td>0.7285</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>rs7518386</td>\n",
       "      <td>chr1:101055617</td>\n",
       "      <td>(T/C)</td>\n",
       "      <td>0.0260</td>\n",
       "      <td>175289</td>\n",
       "      <td>0.7502</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3a</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1219 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        RS_Number           Coord Alleles     MAF  Distance  Dprime      R2  \\\n",
       "0        rs612683  chr1:100880328   (A/T)  0.4004         0  1.0000  1.0000   \n",
       "1      rs12568038  chr1:100879914   (C/T)  0.3415      -414  0.8908  0.6162   \n",
       "2      rs12564838  chr1:100883188   (A/G)  0.3496      2860  0.8609  0.5968   \n",
       "3       rs7517704  chr1:100885419   (A/G)  0.3494      5091  0.8609  0.5962   \n",
       "4      rs12760924  chr1:100898600   (A/T)  0.3546     18272  0.8357  0.5748   \n",
       "...           ...             ...     ...     ...       ...     ...     ...   \n",
       "1214  rs111334083  chr1:100617033   (G/A)  0.0212   -263295  0.8351  0.0101   \n",
       "1215  rs141366089  chr1:100950426   (-/T)  0.0575     70098  0.4970  0.0101   \n",
       "1216   rs10493930  chr1:101079706   (T/C)  0.0833    199378  0.4070  0.0100   \n",
       "1217   rs57411954  chr1:100712556  (AC/-)  0.0276   -167772  0.7285  0.0100   \n",
       "1218    rs7518386  chr1:101055617   (T/C)  0.0260    175289  0.7502  0.0100   \n",
       "\n",
       "     Correlated_Alleles  FORGEdb RegulomeDB Function  \n",
       "0               A=A,T=T      9.0         2b      NaN  \n",
       "1               A=C,T=T      6.0          7      NaN  \n",
       "2               A=A,T=G      6.0          5      NaN  \n",
       "3               A=A,T=G      6.0          6      NaN  \n",
       "4               A=A,T=T      6.0          6      NaN  \n",
       "...                 ...      ...        ...      ...  \n",
       "1214                NaN      7.0          5      NaN  \n",
       "1215                NaN      NaN          .      NaN  \n",
       "1216                NaN      6.0          5      NaN  \n",
       "1217                NaN      NaN          .      NaN  \n",
       "1218                NaN      4.0         3a      NaN  \n",
       "\n",
       "[1219 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Remove the leading single quote\n",
    "data = response.text\n",
    "\n",
    "# Create a StringIO object from the data\n",
    "data_io = io.StringIO(data)\n",
    "\n",
    "# Read the data into a DataFrame using read_csv\n",
    "df = pd.read_csv(data_io, sep='\\\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find overlap b/w 23AndMe and LD\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "not enough values to unpack (expected 2, got 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 15\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m filename \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39mlistdir(directory):\n\u001b[1;32m     12\u001b[0m     \u001b[38;5;66;03m# Check if the file is a CSV file\u001b[39;00m\n\u001b[1;32m     13\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m filename\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m     14\u001b[0m         \u001b[38;5;66;03m# Extract the chromosome number and position from the filename\u001b[39;00m\n\u001b[0;32m---> 15\u001b[0m         chrom, pos \u001b[38;5;241m=\u001b[39m filename\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     16\u001b[0m         pos \u001b[38;5;241m=\u001b[39m pos\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     18\u001b[0m         \u001b[38;5;66;03m# Create a folder for the chromosome number\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: not enough values to unpack (expected 2, got 1)"
     ]
    }
   ],
   "source": [
    "# Separate files in one directory into multiple folders based on chromosome number. The files are named chromosomeNumber_chromosomePosition.csv\n",
    "\n",
    "directory = \"../../Data/ld_variants\"\n",
    "\n",
    "# Create a folder to store the CSV files\n",
    "output_folder = \"../../Data/ld_variants\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Iterate over each file in the directory\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    # Check if the file is a CSV file\n",
    "    if filename.endswith(\".csv\"):\n",
    "        # Extract the chromosome number and position from the filename\n",
    "        chrom, pos = filename.split(\"_\")\n",
    "        pos = pos.split(\".\")[0]\n",
    "        \n",
    "        # Create a folder for the chromosome number\n",
    "        chrom_folder = os.path.join(output_folder, \"chr\" + chrom)\n",
    "        os.makedirs(chrom_folder, exist_ok=True)\n",
    "        \n",
    "        # Move the file to the chromosome folder\n",
    "        src = os.path.join(directory, filename)\n",
    "        dest = os.path.join(chrom_folder, filename)\n",
    "        os.rename(src, dest)\n",
    "        \n",
    "        print(f\"Moved {filename} to {chrom_folder}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 12_14413931.csv\n",
      "Found 18 matching columns\n",
      "Processed 12_115796577.csv\n",
      "Found 43 matching columns\n",
      "Processed 12_115835836.csv\n",
      "Found 53 matching columns\n",
      "Processed 12_120832146.csv\n",
      "Found 70 matching columns\n",
      "Processed 12_103097887.csv\n",
      "Found 102 matching columns\n",
      "Processed 12_96027759.csv\n",
      "Found 112 matching columns\n",
      "Processed 12_85004551.csv\n",
      "Found 171 matching columns\n",
      "Processed 12_111600134.csv\n",
      "Found 239 matching columns\n",
      "Processed 12_28149568.csv\n",
      "Found 272 matching columns\n",
      "Processed 12_29140260.csv\n",
      "Found 310 matching columns\n",
      "Processed 12_57146069.csv\n",
      "Found 337 matching columns\n",
      "Processed 12_293626.csv\n",
      "Found 356 matching columns\n",
      "Processed 12_70798355.csv\n",
      "Found 414 matching columns\n",
      "Processed 12_28174817.csv\n",
      "Found 420 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 12_83064195\n",
      "Added chr12_83064195_G_GA_PRS313_Unknown to matching_columns\n",
      "Processed 12_83064195.csv\n",
      "Found 421 matching columns\n",
      "Processed 12_115108136.csv\n",
      "Found 474 matching columns\n",
      "Processed 12_28347382.csv\n",
      "Found 510 matching columns\n",
      "Processed chromosome 12\n",
      "Found 510 matching columns\n",
      "Found 509 matching variants\n",
      "Processed 15_66630569.csv\n",
      "Found 54 matching columns\n",
      "Processed 15_100905819.csv\n",
      "Found 103 matching columns\n",
      "Processed 15_67457698.csv\n",
      "Found 116 matching columns\n",
      "Processed 15_46680811.csv\n",
      "Found 119 matching columns\n",
      "Processed 15_91512267.csv\n",
      "Found 170 matching columns\n",
      "Processed 15_50694306.csv\n",
      "Found 191 matching columns\n",
      "Processed 15_75750383.csv\n",
      "Found 229 matching columns\n",
      "Processed chromosome 15\n",
      "Found 229 matching columns\n",
      "Found 229 matching variants\n",
      "Processed 14_105213978.csv\n",
      "Found 83 matching columns\n",
      "Processed 14_37228504.csv\n",
      "Found 110 matching columns\n",
      "Processed 14_93070286.csv\n",
      "Found 132 matching columns\n",
      "Processed 14_37128564.csv\n",
      "Found 165 matching columns\n",
      "Processed 14_91841069.csv\n",
      "Found 227 matching columns\n",
      "Processed 14_68660428.csv\n",
      "Found 286 matching columns\n",
      "Processed 14_68979835.csv\n",
      "Found 319 matching columns\n",
      "Processed 14_91751788.csv\n",
      "Found 332 matching columns\n",
      "Processed chromosome 14\n",
      "Found 332 matching columns\n",
      "Found 332 matching variants\n",
      "Processed 13_73806982.csv\n",
      "Found 27 matching columns\n",
      "Processed 13_32839990.csv\n",
      "Found 40 matching columns\n",
      "Processed 13_73960952.csv\n",
      "Found 89 matching columns\n",
      "Processed 13_32972626.csv\n",
      "Found 89 matching columns\n",
      "Processed 13_43501356.csv\n",
      "Found 110 matching columns\n",
      "Processed chromosome 13\n",
      "Found 110 matching columns\n",
      "Found 110 matching variants\n",
      "Processed 22_46283297.csv\n",
      "Found 51 matching columns\n",
      "Processed 22_39343916.csv\n",
      "Found 88 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 22_38583315\n",
      "Added chr22_38583315_AAAAG_AAAAGAAAG,AAAAGAAAGAAAG,A_PRS313_Unknown to matching_columns\n",
      "Processed 22_38583315.csv\n",
      "Found 89 matching columns\n",
      "Processed 22_40904707.csv\n",
      "Found 120 matching columns\n",
      "Processed 22_19766137.csv\n",
      "Found 133 matching columns\n",
      "Processed 22_43433100.csv\n",
      "Found 177 matching columns\n",
      "Processed 22_45319953.csv\n",
      "Found 212 matching columns\n",
      "Processed 22_29203724.csv\n",
      "Found 223 matching columns\n",
      "Processed 22_29135543.csv\n",
      "Found 252 matching columns\n",
      "Processed 22_29121087.csv\n",
      "Found 255 matching columns\n",
      "Processed 22_29551872.csv\n",
      "Found 263 matching columns\n",
      "Processed chromosome 22\n",
      "Found 263 matching columns\n",
      "Found 262 matching variants\n",
      "Processed 4_106069013.csv\n",
      "Found 100 matching columns\n",
      "Processed 4_38784633.csv\n",
      "Found 151 matching columns\n",
      "Processed 4_175842495.csv\n",
      "Found 199 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 4_151218296\n",
      "Added chr4_151218296_CATATTT_C_PRS313_Unknown to matching_columns\n",
      "Processed 4_151218296.csv\n",
      "Found 200 matching columns\n",
      "Processed 4_89240476.csv\n",
      "Found 291 matching columns\n",
      "Processed 4_143467195.csv\n",
      "Found 314 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 4_187503758\n",
      "Added chr4_187503758_A_G,T_PRS313_Unknown to matching_columns\n",
      "Processed 4_187503758.csv\n",
      "Found 315 matching columns\n",
      "Processed 4_175847436.csv\n",
      "Found 330 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 4_126752992\n",
      "Added chr4_126752992_A_AAT,AATAT_PRS313_Unknown to matching_columns\n",
      "Processed 4_126752992.csv\n",
      "Found 331 matching columns\n",
      "Processed 4_92594859.csv\n",
      "Found 371 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 4_84370124\n",
      "Added chr4_84370124_TAA_TA,T_PRS313_Known to matching_columns\n",
      "Processed 4_84370124.csv\n",
      "Found 372 matching columns\n",
      "Processed chromosome 4\n",
      "Found 372 matching columns\n",
      "Found 368 matching variants\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 3_55970777\n",
      "Added chr3_55970777_A_AT_PRS313_Unknown to matching_columns\n",
      "Processed 3_55970777.csv\n",
      "Found 1 matching columns\n",
      "Processed 3_59373745.csv\n",
      "Found 27 matching columns\n",
      "Processed 3_141112859.csv\n",
      "Found 124 matching columns\n",
      "Processed 3_99403877.csv\n",
      "Found 183 matching columns\n",
      "Processed 3_172285237.csv\n",
      "Found 202 matching columns\n",
      "Processed 3_27388664.csv\n",
      "Found 262 matching columns\n",
      "Processed 3_4742251.csv\n",
      "Found 296 matching columns\n",
      "Processed 3_27353716.csv\n",
      "Found 336 matching columns\n",
      "Processed 3_29294845.csv\n",
      "Found 338 matching columns\n",
      "Processed 3_30684907.csv\n",
      "Found 378 matching columns\n",
      "Processed 3_46888198.csv\n",
      "Found 412 matching columns\n",
      "Processed 3_71620370.csv\n",
      "Found 472 matching columns\n",
      "Processed 3_49709912.csv\n",
      "Found 516 matching columns\n",
      "Processed 3_189774456.csv\n",
      "Found 547 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 3_63887449\n",
      "Added chr3_63887449_T_TTG,TTGTG_PRS313_Unknown to matching_columns\n",
      "Processed 3_63887449.csv\n",
      "Found 548 matching columns\n",
      "Processed 3_87037543.csv\n",
      "Found 587 matching columns\n",
      "Processed chromosome 3\n",
      "Found 587 matching columns\n",
      "Found 585 matching variants\n",
      "Processed 2_218292158.csv\n",
      "Found 34 matching columns\n",
      "Processed 2_217920769.csv\n",
      "Found 118 matching columns\n",
      "Processed 2_172974566.csv\n",
      "Found 161 matching columns\n",
      "Processed 2_88358825.csv\n",
      "Found 174 matching columns\n",
      "Processed 2_202204741.csv\n",
      "Found 241 matching columns\n",
      "Processed 2_174212910.csv\n",
      "Found 265 matching columns\n",
      "Processed 2_121246568.csv\n",
      "Found 283 matching columns\n",
      "Processed 2_29615233.csv\n",
      "Found 332 matching columns\n",
      "Processed 2_10138983.csv\n",
      "Found 358 matching columns\n",
      "Processed 2_39699510.csv\n",
      "Found 406 matching columns\n",
      "Processed 2_25129473.csv\n",
      "Found 477 matching columns\n",
      "Processed 2_121058254.csv\n",
      "Found 512 matching columns\n",
      "Processed 2_241388857.csv\n",
      "Found 519 matching columns\n",
      "Processed 2_121159205.csv\n",
      "Found 558 matching columns\n",
      "Processed 2_217955896.csv\n",
      "Found 570 matching columns\n",
      "Processed 2_192381934.csv\n",
      "Found 603 matching columns\n",
      "Processed 2_121089731.csv\n",
      "Found 610 matching columns\n",
      "Processed 2_70172587.csv\n",
      "Found 638 matching columns\n",
      "Processed 2_19315675.csv\n",
      "Found 693 matching columns\n",
      "Processed 2_29179452.csv\n",
      "Found 775 matching columns\n",
      "Processed 2_218714845.csv\n",
      "Found 792 matching columns\n",
      "Processed chromosome 2\n",
      "Found 792 matching columns\n",
      "Found 792 matching variants\n",
      "Processed 5_79180995.csv\n",
      "Found 35 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 5_44508264\n",
      "Added chr5_44508264_G_GT_PRS313_Unknown to matching_columns\n",
      "Processed 5_44508264.csv\n",
      "Found 36 matching columns\n",
      "Processed 5_131640536.csv\n",
      "Found 109 matching columns\n",
      "Processed 5_71965007.csv\n",
      "Found 145 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 5_52679539\n",
      "Added chr5_52679539_C_CA,CAA_PRS313_Unknown to matching_columns\n",
      "Processed 5_52679539.csv\n",
      "Found 146 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 5_32579616\n",
      "Added chr5_32579616_TCA_T_PRS313_Unknown to matching_columns\n",
      "Processed 5_32579616.csv\n",
      "Found 147 matching columns\n",
      "Processed 5_44619502.csv\n",
      "Found 173 matching columns\n",
      "Processed 5_55662540.csv\n",
      "Found 188 matching columns\n",
      "Processed 5_77155397.csv\n",
      "Found 230 matching columns\n",
      "Processed 5_73234583.csv\n",
      "Found 256 matching columns\n",
      "Processed 5_345109.csv\n",
      "Found 328 matching columns\n",
      "Processed 5_122478676.csv\n",
      "Found 363 matching columns\n",
      "Processed 5_2777029.csv\n",
      "Found 399 matching columns\n",
      "Processed 5_56045081.csv\n",
      "Found 437 matching columns\n",
      "Processed 5_122705244.csv\n",
      "Found 461 matching columns\n",
      "Processed 5_1353077.csv\n",
      "Found 465 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 5_81512947\n",
      "Added chr5_81512947_TA_T_PRS313_Unknown to matching_columns\n",
      "Processed 5_81512947.csv\n",
      "Found 466 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 5_176134882\n",
      "Added chr5_176134882_T_A,C_PRS313_Unknown to matching_columns\n",
      "Processed 5_176134882.csv\n",
      "Found 467 matching columns\n",
      "Processed 5_58241712.csv\n",
      "Found 493 matching columns\n",
      "Processed 5_90789470.csv\n",
      "Found 530 matching columns\n",
      "Processed 5_44649944.csv\n",
      "Found 565 matching columns\n",
      "Processed 5_132407058.csv\n",
      "Found 585 matching columns\n",
      "Processed 5_1296255.csv\n",
      "Found 653 matching columns\n",
      "Processed 5_56023083.csv\n",
      "Found 677 matching columns\n",
      "Processed 5_44706498.csv\n",
      "Found 690 matching columns\n",
      "Processed 5_104300273.csv\n",
      "Found 739 matching columns\n",
      "Processed 5_16231194.csv\n",
      "Found 782 matching columns\n",
      "Processed 5_1279790.csv\n",
      "Found 812 matching columns\n",
      "Processed 5_173358154.csv\n",
      "Found 847 matching columns\n",
      "Processed 5_169591460.csv\n",
      "Found 881 matching columns\n",
      "Processed 5_44853593.csv\n",
      "Found 895 matching columns\n",
      "Processed 5_158244083.csv\n",
      "Found 948 matching columns\n",
      "Processed 5_56042972.csv\n",
      "Found 949 matching columns\n",
      "Processed 5_55965167.csv\n",
      "Found 959 matching columns\n",
      "Processed chromosome 5\n",
      "Found 959 matching columns\n",
      "Found 956 matching variants\n",
      "Processed 18_29981526.csv\n",
      "Found 35 matching columns\n",
      "Processed 18_42888797.csv\n",
      "Found 119 matching columns\n",
      "Processed 18_24337424.csv\n",
      "Found 145 matching columns\n",
      "Processed 18_24125857.csv\n",
      "Found 165 matching columns\n",
      "Processed 18_24518050.csv\n",
      "Found 190 matching columns\n",
      "Processed 18_11696613.csv\n",
      "Found 295 matching columns\n",
      "Processed 18_25407513.csv\n",
      "Found 327 matching columns\n",
      "Processed 18_20634253.csv\n",
      "Found 384 matching columns\n",
      "Processed 18_42411803.csv\n",
      "Found 402 matching columns\n",
      "Processed chromosome 18\n",
      "Found 402 matching columns\n",
      "Found 404 matching variants\n",
      "Processed 20_52296849.csv\n",
      "Found 18 matching columns\n",
      "Processed 20_41613706.csv\n",
      "Found 41 matching columns\n",
      "Processed 20_11379842.csv\n",
      "Found 60 matching columns\n",
      "Processed 20_5948227.csv\n",
      "Found 80 matching columns\n",
      "Processed chromosome 20\n",
      "Found 80 matching columns\n",
      "Found 80 matching variants\n",
      "Processed 16_6963972.csv\n",
      "Found 51 matching columns\n",
      "Processed 16_23007047.csv\n",
      "Found 67 matching columns\n",
      "Processed 16_53861592.csv\n",
      "Found 90 matching columns\n",
      "Processed 16_87086492.csv\n",
      "Found 142 matching columns\n",
      "Processed 16_4008542.csv\n",
      "Found 156 matching columns\n",
      "Processed 16_54682064.csv\n",
      "Found 195 matching columns\n",
      "Processed 16_52599188.csv\n",
      "Found 235 matching columns\n",
      "Processed 16_80648296.csv\n",
      "Found 256 matching columns\n",
      "Processed 16_52538825.csv\n",
      "Found 261 matching columns\n",
      "Processed 16_10706580.csv\n",
      "Found 305 matching columns\n",
      "Processed 16_4106788.csv\n",
      "Found 324 matching columns\n",
      "Processed 16_85145977.csv\n",
      "Found 390 matching columns\n",
      "Processed 16_53809123.csv\n",
      "Found 452 matching columns\n",
      "Processed 16_53861139.csv\n",
      "Found 455 matching columns\n",
      "Processed chromosome 16\n",
      "Found 455 matching columns\n",
      "Found 455 matching variants\n",
      "Processed 11_111696440.csv\n",
      "Found 26 matching columns\n",
      "Processed 11_122966626.csv\n",
      "Found 100 matching columns\n",
      "Processed 11_69331418.csv\n",
      "Found 132 matching columns\n",
      "Processed 11_1895708.csv\n",
      "Found 217 matching columns\n",
      "Processed 11_803017.csv\n",
      "Found 278 matching columns\n",
      "Processed 11_433617.csv\n",
      "Found 305 matching columns\n",
      "Processed 11_46318032.csv\n",
      "Found 324 matching columns\n",
      "Processed 11_42844441.csv\n",
      "Found 359 matching columns\n",
      "Processed 11_69330983.csv\n",
      "Found 365 matching columns\n",
      "Processed 11_65572431.csv\n",
      "Found 427 matching columns\n",
      "Processed 11_108267402.csv\n",
      "Found 480 matching columns\n",
      "Processed 11_116727936.csv\n",
      "Found 608 matching columns\n",
      "Processed 11_129243417.csv\n",
      "Found 630 matching columns\n",
      "Processed 11_129461016.csv\n",
      "Found 649 matching columns\n",
      "Processed 11_103614438.csv\n",
      "Found 703 matching columns\n",
      "Processed 11_18664241.csv\n",
      "Found 794 matching columns\n",
      "Processed 11_69328130.csv\n",
      "Found 832 matching columns\n",
      "Processed 11_65553492.csv\n",
      "Found 860 matching columns\n",
      "Processed 11_44368892.csv\n",
      "Found 906 matching columns\n",
      "Processed chromosome 11\n",
      "Found 906 matching columns\n",
      "Found 906 matching variants\n",
      "Processed 10_38523626.csv\n",
      "Found 28 matching columns\n",
      "Processed 10_123340107.csv\n",
      "Found 60 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 10_22861490\n",
      "Added chr10_22861490_A_C,T_PRS313_Unknown to matching_columns\n",
      "Processed 10_22861490.csv\n",
      "Found 61 matching columns\n",
      "Processed 10_123349324.csv\n",
      "Found 68 matching columns\n",
      "Processed 10_123095209.csv\n",
      "Found 83 matching columns\n",
      "Processed 10_13892298.csv\n",
      "Found 108 matching columns\n",
      "Processed 10_5794652.csv\n",
      "Found 145 matching columns\n",
      "Processed 10_115128491.csv\n",
      "Found 242 matching columns\n",
      "Processed 10_22032942.csv\n",
      "Found 315 matching columns\n",
      "Processed 10_64299890.csv\n",
      "Found 351 matching columns\n",
      "Processed 10_80886726.csv\n",
      "Found 374 matching columns\n",
      "Processed 10_123340431.csv\n",
      "Found 389 matching columns\n",
      "Processed 10_80851257.csv\n",
      "Found 406 matching columns\n",
      "Processed 10_22477776.csv\n",
      "Found 435 matching columns\n",
      "Processed 10_95292187.csv\n",
      "Found 453 matching columns\n",
      "Processed 10_71335574.csv\n",
      "Found 478 matching columns\n",
      "Processed 10_64819996.csv\n",
      "Found 519 matching columns\n",
      "Processed 10_114777670.csv\n",
      "Found 592 matching columns\n",
      "Processed chromosome 10\n",
      "Found 592 matching columns\n",
      "Found 591 matching variants\n",
      "Processed 17_43212339.csv\n",
      "Found 39 matching columns\n",
      "Processed 17_53209774.csv\n",
      "Found 73 matching columns\n",
      "Processed 17_44283858.csv\n",
      "Found 127 matching columns\n",
      "Processed 17_40485239.csv\n",
      "Found 148 matching columns\n",
      "Processed 17_77781725.csv\n",
      "Found 197 matching columns\n",
      "Processed 17_40127060.csv\n",
      "Found 205 matching columns\n",
      "Processed 17_29168077.csv\n",
      "Found 242 matching columns\n",
      "Processed 17_39251123.csv\n",
      "Found 254 matching columns\n",
      "Processed 17_40744470.csv\n",
      "Found 259 matching columns\n",
      "Processed chromosome 17\n",
      "Found 259 matching columns\n",
      "Found 261 matching variants\n",
      "Processed 21_16566350.csv\n",
      "Found 26 matching columns\n",
      "Processed 21_47762932.csv\n",
      "Found 43 matching columns\n",
      "Processed 21_16364756.csv\n",
      "Found 65 matching columns\n",
      "Processed 21_16574455.csv\n",
      "Found 87 matching columns\n",
      "Processed chromosome 21\n",
      "Found 87 matching columns\n",
      "Found 87 matching variants\n",
      "Processed 19_18569492.csv\n",
      "Found 58 matching columns\n",
      "Processed 19_46166073.csv\n",
      "Found 87 matching columns\n",
      "Processed 19_44283031.csv\n",
      "Found 128 matching columns\n",
      "Processed 19_13249921.csv\n",
      "Found 145 matching columns\n",
      "Processed 19_19517054.csv\n",
      "Found 199 matching columns\n",
      "Processed 19_17393925.csv\n",
      "Found 240 matching columns\n",
      "Processed 19_55816678.csv\n",
      "Found 280 matching columns\n",
      "Processed chromosome 19\n",
      "Found 280 matching columns\n",
      "Found 280 matching variants\n",
      "Processed 7_25569548.csv\n",
      "Found 43 matching columns\n",
      "SNP not found in dbSNP and cannot be proxied using LDProxy 7_91459189\n",
      "Added chr7_91459189_A_AT,ATT_PRS313_Unknown to matching_columns\n",
      "Processed 7_91459189.csv\n",
      "Found 44 matching columns\n",
      "Processed 7_98005235.csv\n",
      "Found 82 matching columns\n",
      "Processed 7_28869017.csv\n",
      "Found 124 matching columns\n",
      "Processed 7_55192256.csv\n",
      "Found 161 matching columns\n",
      "Processed 7_144048902.csv\n",
      "Found 194 matching columns\n",
      "Processed 7_101552440.csv\n",
      "Found 213 matching columns\n",
      "Processed 7_99948655.csv\n",
      "Found 252 matching columns\n",
      "Processed 7_139943702.csv\n",
      "Found 264 matching columns\n",
      "Processed 7_94113799.csv\n",
      "Found 322 matching columns\n",
      "Processed 7_130674481.csv\n",
      "Found 346 matching columns\n",
      "Processed 7_21940960.csv\n",
      "Found 404 matching columns\n",
      "Processed 7_102481842.csv\n",
      "Found 438 matching columns\n",
      "Processed 7_130656911.csv\n",
      "Found 444 matching columns\n",
      "Processed chromosome 7\n",
      "Found 444 matching columns\n",
      "Found 443 matching variants\n",
      "Processed 9_98362587.csv\n",
      "Found 38 matching columns\n",
      "Processed 9_6880263.csv\n",
      "Found 109 matching columns\n",
      "Processed 9_110885479.csv\n",
      "Found 177 matching columns\n",
      "Processed 9_110837176.csv\n",
      "Found 195 matching columns\n",
      "Processed 9_21964882.csv\n",
      "Found 247 matching columns\n",
      "Processed 9_36928288.csv\n",
      "Found 272 matching columns\n",
      "Processed 9_110303808.csv\n",
      "Found 283 matching columns\n",
      "Processed 9_129424719.csv\n",
      "Found 307 matching columns\n",
      "Processed 9_119313486.csv\n",
      "Found 360 matching columns\n",
      "Processed 9_110849525.csv\n",
      "Found 404 matching columns\n",
      "Processed 9_136146597.csv\n",
      "Found 476 matching columns\n",
      "Processed 9_87782211.csv\n",
      "Found 506 matching columns\n",
      "Processed 9_22041998.csv\n",
      "Found 518 matching columns\n",
      "Processed 9_110837073.csv\n",
      "Found 533 matching columns\n",
      "Processed chromosome 9\n",
      "Found 533 matching columns\n",
      "Found 533 matching variants\n",
      "Processed 8_106358620.csv\n",
      "Found 29 matching columns\n",
      "Processed 8_143669254.csv\n",
      "Found 78 matching columns\n",
      "Processed 8_76333056.csv\n",
      "Found 98 matching columns\n",
      "Processed 8_124571581.csv\n",
      "Found 122 matching columns\n",
      "Processed 8_170692.csv\n",
      "Found 145 matching columns\n",
      "Processed 8_76230943.csv\n",
      "Found 165 matching columns\n",
      "Processed 8_117209548.csv\n",
      "Found 231 matching columns\n",
      "Processed 8_128213561.csv\n",
      "Found 277 matching columns\n",
      "Processed 8_124739913.csv\n",
      "Found 294 matching columns\n",
      "Processed 8_29509616.csv\n",
      "Found 318 matching columns\n",
      "Processed 8_128372172.csv\n",
      "Found 344 matching columns\n",
      "Processed 8_23447496.csv\n",
      "Found 379 matching columns\n",
      "Processed 8_102483100.csv\n",
      "Found 438 matching columns\n",
      "Processed 8_120862186.csv\n",
      "Found 473 matching columns\n",
      "Processed 8_124563705.csv\n",
      "Found 478 matching columns\n",
      "Processed 8_76378165.csv\n",
      "Found 499 matching columns\n",
      "Processed 8_128370949.csv\n",
      "Found 505 matching columns\n",
      "Processed 8_36858483.csv\n",
      "Found 529 matching columns\n",
      "Processed 8_17787610.csv\n",
      "Found 578 matching columns\n",
      "Processed 8_23663653.csv\n",
      "Found 610 matching columns\n",
      "Processed 8_129199566.csv\n",
      "Found 652 matching columns\n",
      "Processed chromosome 8\n",
      "Found 652 matching columns\n",
      "Found 652 matching variants\n",
      "Processed 6_152432902.csv\n",
      "Found 18 matching columns\n",
      "Processed 6_152055978.csv\n",
      "Found 42 matching columns\n",
      "Processed 6_16399557.csv\n",
      "Found 56 matching columns\n",
      "Processed 6_152023191.csv\n",
      "Found 103 matching columns\n",
      "Processed 6_82263549.csv\n",
      "Found 141 matching columns\n",
      "Processed 6_152022664.csv\n",
      "Found 158 matching columns\n",
      "Processed 6_18783140.csv\n",
      "Found 201 matching columns\n",
      "Processed 6_87803819.csv\n",
      "Found 257 matching columns\n",
      "Processed 6_20537845.csv\n",
      "Found 324 matching columns\n",
      "Processed 6_85912194.csv\n",
      "Found 345 matching columns\n",
      "Processed 6_151949806.csv\n",
      "Found 366 matching columns\n",
      "Processed 6_169006947.csv\n",
      "Found 401 matching columns\n",
      "Processed 6_13713366.csv\n",
      "Found 429 matching columns\n",
      "Processed 6_21923810.csv\n",
      "Found 454 matching columns\n",
      "Processed 6_43227141.csv\n",
      "Found 488 matching columns\n",
      "Processed 6_151955914.csv\n",
      "Found 502 matching columns\n",
      "Processed 6_170332621.csv\n",
      "Found 553 matching columns\n",
      "Processed 6_149595505.csv\n",
      "Found 601 matching columns\n",
      "Processed 6_27425644.csv\n",
      "Found 629 matching columns\n",
      "Processed 6_130341728.csv\n",
      "Found 670 matching columns\n",
      "Processed chromosome 6\n",
      "Found 670 matching columns\n",
      "Found 670 matching variants\n",
      "Processed 1_88156923.csv\n",
      "Found 24 matching columns\n",
      "Processed 1_120257110.csv\n",
      "Found 62 matching columns\n",
      "Processed 1_114445880.csv\n",
      "Found 116 matching columns\n",
      "Processed 1_88428199.csv\n",
      "Found 131 matching columns\n",
      "Processed 1_202184600.csv\n",
      "Found 161 matching columns\n",
      "Processed 1_121280613.csv\n",
      "Found 172 matching columns\n",
      "Processed 1_118141492.csv\n",
      "Found 216 matching columns\n",
      "Processed 1_168171052.csv\n",
      "Found 248 matching columns\n",
      "Processed 1_110198129.csv\n",
      "Found 342 matching columns\n",
      "Processed 1_220671050.csv\n",
      "Found 382 matching columns\n",
      "Processed 1_46670206.csv\n",
      "Found 443 matching columns\n",
      "Processed 1_41389220.csv\n",
      "Found 449 matching columns\n",
      "Processed 1_208076291.csv\n",
      "Found 534 matching columns\n",
      "Processed 1_51467096.csv\n",
      "Found 563 matching columns\n",
      "Processed 1_100880328.csv\n",
      "Found 596 matching columns\n",
      "Processed 1_149906413.csv\n",
      "Found 618 matching columns\n",
      "Processed 1_18807339.csv\n",
      "Found 658 matching columns\n",
      "Processed 1_7917076.csv\n",
      "Found 708 matching columns\n",
      "Processed 1_217053815.csv\n",
      "Found 790 matching columns\n",
      "Processed 1_145604302.csv\n",
      "Found 816 matching columns\n",
      "Processed 1_203770448.csv\n",
      "Found 842 matching columns\n",
      "Processed 1_10566215.csv\n",
      "Found 910 matching columns\n",
      "Processed 1_242034263.csv\n",
      "Found 917 matching columns\n",
      "Processed 1_41380440.csv\n",
      "Found 943 matching columns\n",
      "Processed 1_217220574.csv\n",
      "Found 961 matching columns\n",
      "Processed 1_155556971.csv\n",
      "Found 1010 matching columns\n",
      "Processed 1_201437832.csv\n",
      "Found 1025 matching columns\n",
      "Processed 1_204502514.csv\n",
      "Found 1092 matching columns\n",
      "Processed 1_172328767.csv\n",
      "Found 1130 matching columns\n",
      "Processed 1_121287994.csv\n",
      "Found 1132 matching columns\n",
      "Processed chromosome 1\n",
      "Found 1132 matching columns\n",
      "Found 1132 matching variants\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "Cannot save file into a non-existent directory: '../Final_training_data'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 99\u001b[0m\n\u001b[1;32m     96\u001b[0m matching_variants_all_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat(matching_variants_all, ignore_index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     98\u001b[0m \u001b[38;5;66;03m# Save the dataframe with matching variants across all chromosomes\u001b[39;00m\n\u001b[0;32m---> 99\u001b[0m \u001b[43mmatching_variants_all_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../Final_training_data/23AndMe_matching_variants.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/generic.py:3902\u001b[0m, in \u001b[0;36mNDFrame.to_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[1;32m   3891\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[1;32m   3893\u001b[0m formatter \u001b[38;5;241m=\u001b[39m DataFrameFormatter(\n\u001b[1;32m   3894\u001b[0m     frame\u001b[38;5;241m=\u001b[39mdf,\n\u001b[1;32m   3895\u001b[0m     header\u001b[38;5;241m=\u001b[39mheader,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3899\u001b[0m     decimal\u001b[38;5;241m=\u001b[39mdecimal,\n\u001b[1;32m   3900\u001b[0m )\n\u001b[0;32m-> 3902\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3903\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3904\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3905\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3906\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3907\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3908\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3909\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3910\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3911\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3912\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3913\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3914\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/io/formats/format.py:1152\u001b[0m, in \u001b[0;36mDataFrameRenderer.to_csv\u001b[0;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[1;32m   1131\u001b[0m     created_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m   1133\u001b[0m csv_formatter \u001b[38;5;241m=\u001b[39m CSVFormatter(\n\u001b[1;32m   1134\u001b[0m     path_or_buf\u001b[38;5;241m=\u001b[39mpath_or_buf,\n\u001b[1;32m   1135\u001b[0m     lineterminator\u001b[38;5;241m=\u001b[39mlineterminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1150\u001b[0m     formatter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfmt,\n\u001b[1;32m   1151\u001b[0m )\n\u001b[0;32m-> 1152\u001b[0m \u001b[43mcsv_formatter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1154\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[1;32m   1155\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/io/formats/csvs.py:247\u001b[0m, in \u001b[0;36mCSVFormatter.save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    243\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    244\u001b[0m \u001b[38;5;124;03mCreate the writer & save.\u001b[39;00m\n\u001b[1;32m    245\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    246\u001b[0m \u001b[38;5;66;03m# apply compression and byte/text conversion\u001b[39;00m\n\u001b[0;32m--> 247\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    249\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    250\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    251\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    252\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    254\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[1;32m    255\u001b[0m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[1;32m    256\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwriter \u001b[38;5;241m=\u001b[39m csvlib\u001b[38;5;241m.\u001b[39mwriter(\n\u001b[1;32m    257\u001b[0m         handles\u001b[38;5;241m.\u001b[39mhandle,\n\u001b[1;32m    258\u001b[0m         lineterminator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlineterminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    263\u001b[0m         quotechar\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquotechar,\n\u001b[1;32m    264\u001b[0m     )\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_save()\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/io/common.py:739\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    737\u001b[0m \u001b[38;5;66;03m# Only for write methods\u001b[39;00m\n\u001b[1;32m    738\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode \u001b[38;5;129;01mand\u001b[39;00m is_path:\n\u001b[0;32m--> 739\u001b[0m     \u001b[43mcheck_parent_directory\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    741\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m compression:\n\u001b[1;32m    742\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m compression \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mzstd\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    743\u001b[0m         \u001b[38;5;66;03m# compression libraries do not like an explicit text-mode\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/io/common.py:604\u001b[0m, in \u001b[0;36mcheck_parent_directory\u001b[0;34m(path)\u001b[0m\n\u001b[1;32m    602\u001b[0m parent \u001b[38;5;241m=\u001b[39m Path(path)\u001b[38;5;241m.\u001b[39mparent\n\u001b[1;32m    603\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m parent\u001b[38;5;241m.\u001b[39mis_dir():\n\u001b[0;32m--> 604\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\u001b[38;5;124mrf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot save file into a non-existent directory: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mparent\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mOSError\u001b[0m: Cannot save file into a non-existent directory: '../Final_training_data'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "directory = \"../../Data/ld_variants\"\n",
    "\n",
    "# Create a folder to store the CSV files\n",
    "output_folder = \"../../Data/ld_variants\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "\n",
    "def extract_coord_alleles(col_name):\n",
    "    match = re.search(r'chr(\\d+)_(\\d+)_([ACGT]+)_([ACGT,]+)', col_name)\n",
    "    if match:\n",
    "        chr_num, position, ref_allele, alt_alleles = match.groups()\n",
    "        return f'chr{chr_num}:{position}', f'({ref_allele}/{alt_alleles})'\n",
    "    return None, None\n",
    "\n",
    "# Create an empty list to store the matching variants across all chromosomes\n",
    "matching_variants_all = []\n",
    "\n",
    "# Iterate over each chromosome folder\n",
    "for chrom_folder in os.listdir(output_folder):\n",
    "    # Check if the folder is a chromosome folder\n",
    "    if chrom_folder.startswith(\"chr\"):\n",
    "        # Extract the chromosome number\n",
    "        chrom = chrom_folder[3:]\n",
    "\n",
    "        files_folder = os.path.join(output_folder, chrom_folder)\n",
    "        \n",
    "        # Load the training data for the chromosome\n",
    "        training_data = pd.read_parquet(f\"../../Data/Raw_training_data/23AndMe_PRS313_merged_chr{chrom}.parquet\")\n",
    "        \n",
    "        # Create empty lists to store the matching columns and variants for the chromosome\n",
    "        matching_columns = []\n",
    "        matching_variants_chrom = []\n",
    "        not_found_snps_chrom = []\n",
    "\n",
    "        # Read all the files in the folder\n",
    "        for filename in os.listdir(files_folder):\n",
    "\n",
    "            # Check if the file is a CSV file\n",
    "            if filename.endswith(\".csv\"):\n",
    "                # Load the variants for the file\n",
    "                variants = pd.read_csv(os.path.join(files_folder, filename))\n",
    "                \n",
    "                # Find the matching columns in the training data\n",
    "                for column in training_data.columns:\n",
    "                    coord, alleles = extract_coord_alleles(column)\n",
    "                    if coord is not None and alleles is not None:\n",
    "                        try:\n",
    "                            # Got rid of allele matching, because of coding differences. GC--> C (in PRS313) is coded as (C/-) in dbSNP and LDProxy\n",
    "                            mask = (variants['Coord'] == coord)\n",
    "                            # mask = (variants['Coord'] == coord) & (variants['Alleles'] == alleles)\n",
    "                                \n",
    "                            if mask.any():\n",
    "                                # Only append column if it has not been added before\n",
    "                                if column not in matching_columns:\n",
    "                                    matching_columns.append(column)\n",
    "                                    matching_variants_chrom.append(variants[mask])\n",
    "                        except KeyError:\n",
    "                            error_position = filename.split('.')[0]\n",
    "                            not_found_snps_chrom.append(error_position)\n",
    "                            print(\"SNP not found in dbSNP and cannot be proxied using LDProxy\", error_position)\n",
    "\n",
    "                            # Find the column in training_data.columns that contains the error_position\n",
    "                            for col in training_data.columns:\n",
    "                                if error_position in col:\n",
    "                                    # Just to ensure that the PRS313 column is added even if the LDProxy information is missing\n",
    "                                    matching_columns.append(col)\n",
    "                                    print(f\"Added {col} to matching_columns\")\n",
    "                                    break\n",
    "                            break\n",
    "\n",
    "                print(f\"Processed {filename}\")\n",
    "                print(f\"Found {len(matching_columns)} matching columns\")\n",
    "                \n",
    "\n",
    "        # Get the matching columns from the training data\n",
    "        matching_data = training_data[matching_columns]\n",
    "\n",
    "        # Save the matching data for the chromosome\n",
    "        matching_data.to_parquet(f\"../../Data/Filtered_training_data/23AndMe_PRS313_merged_chr{chrom}_matching.parquet\")\n",
    "        \n",
    "        # Concatenate the matching variants for the chromosome into a single dataframe\n",
    "        matching_variants_chrom_df = pd.concat(matching_variants_chrom, ignore_index=True)\n",
    "        \n",
    "        # Append the matching variants for the chromosome to the overall list\n",
    "        matching_variants_all.append(matching_variants_chrom_df)\n",
    "\n",
    "        print(f\"Processed chromosome {chrom}\")\n",
    "        print(f\"Found {len(matching_columns)} matching columns\")\n",
    "        print(f\"Found {len(matching_variants_chrom_df)} matching variants\")\n",
    "\n",
    "# Concatenate the matching variants from all chromosomes into a single dataframe\n",
    "matching_variants_all_df = pd.concat(matching_variants_all, ignore_index=True)\n",
    "\n",
    "# Save the dataframe with matching variants across all chromosomes\n",
    "matching_variants_all_df.to_csv(\"../../Data/Filtered_training_data/23AndMe_matching_variants.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../positions/Missing_ld.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 17\u001b[0m\n\u001b[1;32m     14\u001b[0m r2_threshold \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.01\u001b[39m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# Read the RSIDs from the text file\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../positions/Missing_ld.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m file:\n\u001b[1;32m     18\u001b[0m     rsids \u001b[38;5;241m=\u001b[39m file\u001b[38;5;241m.\u001b[39mread()\u001b[38;5;241m.\u001b[39mstrip()\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     20\u001b[0m \u001b[38;5;66;03m# Create a folder to store the CSV files\u001b[39;00m\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/IPython/core/interactiveshell.py:284\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    277\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[1;32m    278\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    279\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    280\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    282\u001b[0m     )\n\u001b[0;32m--> 284\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mio_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../positions/Missing_ld.txt'"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import os\n",
    "import io\n",
    "\n",
    "# API key\n",
    "api_key = \"ac16be4ad92d\"\n",
    "\n",
    "# Base URL for LDProxy API\n",
    "base_url = \"https://ldlink.nih.gov/LDlinkRest/ldproxy\"\n",
    "\n",
    "population = \"ALL\"\n",
    "window = 1000000\n",
    "r2_threshold = 0.01\n",
    "\n",
    "# Read the RSIDs from the text file\n",
    "with open(\"../positions/Missing_ld.txt\", \"r\") as file:\n",
    "    rsids = file.read().strip().split(\"\\n\")\n",
    "\n",
    "# Create a folder to store the CSV files\n",
    "output_folder = \"../ld_variants\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Iterate over each RSID\n",
    "for rsid in rsids:\n",
    "    # Construct the API request URL\n",
    "    url = f\"{base_url}?var={rsid}&pop={population}&r2_d=r2&window={window}&genome_build=grch37&token={api_key}\"\n",
    "\n",
    "    # Send the API request\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        # Create a StringIO object from the data\n",
    "        data_io = io.StringIO(response.text)\n",
    "\n",
    "        # Read the data into a DataFrame using read_csv\n",
    "        df = pd.read_csv(data_io, sep='\\\\t')\n",
    "\n",
    "        # Generate a unique filename for the CSV file\n",
    "        output_file = os.path.join(output_folder, f\"{rsid}.csv\")\n",
    "\n",
    "        # Save the high LD variants to a CSV file\n",
    "        df.to_csv(output_file, index=False)\n",
    "\n",
    "        print(f\"Saved high LD variants for {rsid} to {output_file}\")\n",
    "    else:\n",
    "        print(f\"Failed to retrieve data for {rsid}. Status code: {response.status_code}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
