{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "def read_vcf_header(file_path):\n",
    "    \"\"\"Reads the VCF file header (column names) and returns it.\"\"\"\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('#CHROM'):\n",
    "                return line.strip().split('\\t')\n",
    "    return []\n",
    "\n",
    "def extract_snp_data(file_path, header):\n",
    "    \"\"\"Extracts SNP data from a VCF file.\"\"\"\n",
    "    snp_data = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('#'):\n",
    "                fields = line.strip().split('\\t')\n",
    "                chrom, pos, ref, alt = fields[0], fields[1], fields[3], fields[4]\n",
    "                snp_id = f\"chr{chrom}_{pos}_{ref}_{alt}\"\n",
    "                genotype_data = fields[9:]  # Genotype data starts from the 10th column\n",
    "                snp_data.append((snp_id, genotype_data))\n",
    "    return snp_data\n",
    "\n",
    "def process_vcf_files(directory):\n",
    "    \"\"\"Processes all VCF files in the specified directory.\"\"\"\n",
    "    vcf_files = glob.glob(os.path.join(directory, '*.vcf'))\n",
    "    all_snp_data = []\n",
    "    patient_ids = None\n",
    "    \n",
    "    for file_path in vcf_files:\n",
    "        if patient_ids is None:\n",
    "            patient_ids = read_vcf_header(file_path)[9:]  # Assuming the first 9 columns are standard VCF columns\n",
    "        snp_data = extract_snp_data(file_path, patient_ids)\n",
    "\n",
    "        if (len(snp_data) == 0):\n",
    "            # print(\"No matches found at \", file_path)\n",
    "            continue\n",
    "        else:\n",
    "            for SNP in snp_data:\n",
    "                position_in_vcf = SNP[0].split(\"_\")[1]\n",
    "                file_path_position = file_path.split(\"_pos\")[1][:-4]\n",
    "                if (position_in_vcf == file_path_position):\n",
    "                    all_snp_data.extend(snp_data)\n",
    "                    # print(\"Adding\")\n",
    "                else:\n",
    "                    # Ignoring copy number variations\n",
    "                    # print(\"Ignoring\")\n",
    "                    continue\n",
    "\n",
    "    # Convert the collected SNP data to a DataFrame\n",
    "    data_dict = {snp_id: genotypes for snp_id, genotypes in all_snp_data}\n",
    "    df = pd.DataFrame(data_dict, index=patient_ids)\n",
    "    print(df.shape)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find intersection b/w 23AndMe and PRS313"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated PRS313_df:\n",
      "           chr2_192381934_C_T chr8_129199566_G_A chr8_29509616_A_C  \\\n",
      "HG00096                   1|1                0|1               1|1   \n",
      "HG00097                   1|1                1|0               1|1   \n",
      "HG00099                   1|1                0|0               0|0   \n",
      "HG00100                   0|1                0|0               0|1   \n",
      "HG00101                   1|1                0|0               1|1   \n",
      "...                       ...                ...               ...   \n",
      "NA21141                   1|1                0|0               0|0   \n",
      "NA21142                   1|0                1|0               0|1   \n",
      "NA21143                   1|1                0|0               1|1   \n",
      "NA21144                   1|1                0|0               0|1   \n",
      "in_23andMe               True              False              True   \n",
      "\n",
      "           chr2_121089731_T_C chr15_75750383_T_C chr18_24337424_C_G  \\\n",
      "HG00096                   1|0                1|0                1|1   \n",
      "HG00097                   0|0                0|0                0|1   \n",
      "HG00099                   0|1                0|0                1|1   \n",
      "HG00100                   0|0                0|0                0|1   \n",
      "HG00101                   1|0                1|1                1|1   \n",
      "...                       ...                ...                ...   \n",
      "NA21141                   0|0                0|1                1|1   \n",
      "NA21142                   0|0                1|0                0|1   \n",
      "NA21143                   0|0                1|1                0|1   \n",
      "NA21144                   0|0                0|1                1|0   \n",
      "in_23andMe               True              False               True   \n",
      "\n",
      "           chr5_131640536_A_G chr16_4008542_CAAAAA_C chr16_10706580_G_A  \\\n",
      "HG00096                   0|1                    1|1                0|0   \n",
      "HG00097                   1|1                    1|1                0|0   \n",
      "HG00099                   0|1                    1|1                0|0   \n",
      "HG00100                   1|0                    1|1                0|0   \n",
      "HG00101                   1|1                    1|1                0|0   \n",
      "...                       ...                    ...                ...   \n",
      "NA21141                   1|1                    1|1                0|0   \n",
      "NA21142                   0|1                    1|1                1|0   \n",
      "NA21143                   0|1                    1|1                0|0   \n",
      "NA21144                   0|0                    0|1                0|1   \n",
      "in_23andMe              False                  False              False   \n",
      "\n",
      "           chr9_110885479_C_T  ... chr12_120832146_C_T  \\\n",
      "HG00096                   1|1  ...                 0|0   \n",
      "HG00097                   1|0  ...                 1|0   \n",
      "HG00099                   1|0  ...                 0|0   \n",
      "HG00100                   1|1  ...                 0|1   \n",
      "HG00101                   0|1  ...                 0|1   \n",
      "...                       ...  ...                 ...   \n",
      "NA21141                   1|1  ...                 0|0   \n",
      "NA21142                   1|1  ...                 1|0   \n",
      "NA21143                   1|1  ...                 0|0   \n",
      "NA21144                   1|1  ...                 1|0   \n",
      "in_23andMe              False  ...                True   \n",
      "\n",
      "           chr3_63887449_T_TTG,TTGTG chr4_126752992_A_AAT,AATAT  \\\n",
      "HG00096                          0|0                        0|1   \n",
      "HG00097                          0|1                        1|1   \n",
      "HG00099                          0|0                        0|1   \n",
      "HG00100                          1|0                        1|0   \n",
      "HG00101                          0|1                        1|0   \n",
      "...                              ...                        ...   \n",
      "NA21141                          0|0                        1|1   \n",
      "NA21142                          0|1                        0|0   \n",
      "NA21143                          0|0                        1|0   \n",
      "NA21144                          0|0                        0|1   \n",
      "in_23andMe                     False                      False   \n",
      "\n",
      "           chr4_187503758_A_G,T chr4_84370124_TAA_TA,T chr5_176134882_T_A,C  \\\n",
      "HG00096                     0|1                    0|1                  0|1   \n",
      "HG00097                     0|1                    1|0                  1|0   \n",
      "HG00099                     1|1                    0|0                  0|0   \n",
      "HG00100                     1|0                    1|1                  1|1   \n",
      "HG00101                     1|0                    1|0                  1|0   \n",
      "...                         ...                    ...                  ...   \n",
      "NA21141                     0|0                    1|0                  1|1   \n",
      "NA21142                     0|1                    0|1                  1|1   \n",
      "NA21143                     0|1                    1|1                  0|0   \n",
      "NA21144                     0|0                    1|0                  1|1   \n",
      "in_23andMe                False                  False                False   \n",
      "\n",
      "           chr5_52679539_C_CA,CAA chr7_91459189_A_AT,ATT chr10_22861490_A_C,T  \\\n",
      "HG00096                       0|0                    1|0                  1|1   \n",
      "HG00097                       0|0                    1|0                  1|1   \n",
      "HG00099                       0|0                    1|1                  1|1   \n",
      "HG00100                       0|0                    0|1                  0|1   \n",
      "HG00101                       0|0                    0|0                  1|1   \n",
      "...                           ...                    ...                  ...   \n",
      "NA21141                       0|0                    1|1                  1|1   \n",
      "NA21142                       0|0                    0|1                  1|1   \n",
      "NA21143                       1|0                    1|0                  0|1   \n",
      "NA21144                       0|0                    1|0                  1|1   \n",
      "in_23andMe                  False                  False                False   \n",
      "\n",
      "           chr22_38583315_AAAAG_AAAAGAAAG,AAAAGAAAGAAAG,A  \n",
      "HG00096                                               0|0  \n",
      "HG00097                                               0|0  \n",
      "HG00099                                               1|1  \n",
      "HG00100                                               0|0  \n",
      "HG00101                                               0|0  \n",
      "...                                                   ...  \n",
      "NA21141                                               0|1  \n",
      "NA21142                                               0|1  \n",
      "NA21143                                               1|0  \n",
      "NA21144                                               1|0  \n",
      "in_23andMe                                          False  \n",
      "\n",
      "[2505 rows x 313 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define file paths\n",
    "ttAndMeFullPanel = \"../../../Data/23andMe_metadata_files/23andMeGenePanel.csv\"\n",
    "prs313_file = './concatenated_snps_processed.csv'\n",
    "\n",
    "# Read the CSV files into DataFrames\n",
    "ttAndMe = pd.read_csv(ttAndMeFullPanel)\n",
    "PRS313_df = pd.read_csv(prs313_file, index_col=0)\n",
    "\n",
    "# Extract chromosome and position information from PRS313_df columns\n",
    "PRS313_info = PRS313_df.columns.str.extract(r'chr(\\d+)_(\\d+)', expand=True)\n",
    "PRS313_info.columns = ['chromosome', 'position']\n",
    "PRS313_info['full_col'] = PRS313_df.columns\n",
    "\n",
    "# Convert chromosome and position columns to string type for comparison\n",
    "ttAndMe['chromosome'] = ttAndMe['chromosome'].astype(str)\n",
    "ttAndMe['position'] = ttAndMe['position'].astype(str)\n",
    "PRS313_info['chromosome'] = PRS313_info['chromosome'].astype(str)\n",
    "PRS313_info['position'] = PRS313_info['position'].astype(str)\n",
    "\n",
    "# Create a new row in PRS313_df to store the boolean values\n",
    "PRS313_df.loc['in_23andMe'] = False\n",
    "\n",
    "# Merge ttAndMe with PRS313_info to identify matching rows\n",
    "merged_df = pd.merge(ttAndMe, PRS313_info, how='inner', on=['chromosome', 'position'])\n",
    "\n",
    "# Set the 'in_23andMe' row to True for matching columns\n",
    "PRS313_df.loc['in_23andMe', merged_df['full_col']] = True\n",
    "\n",
    "# Print the updated PRS313_df DataFrame\n",
    "print(\"Updated PRS313_df:\")\n",
    "print(PRS313_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HG00096         1|1\n",
       "HG00097         0|0\n",
       "HG00099         1|1\n",
       "HG00100         1|0\n",
       "HG00101         1|1\n",
       "              ...  \n",
       "NA21141         0|0\n",
       "NA21142         1|0\n",
       "NA21143         0|0\n",
       "NA21144         0|0\n",
       "in_23andMe    False\n",
       "Name: chr22_45319953_G_A, Length: 2505, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the max value in PRS313_df\n",
    "\n",
    "PRS313_df.iloc[:,23]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 53891)\n",
      "10 columns renamed with suffix '_PRS313_Known'\n",
      "20 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 53134)\n",
      "4 columns renamed with suffix '_PRS313_Known'\n",
      "17 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 44862)\n",
      "3 columns renamed with suffix '_PRS313_Known'\n",
      "13 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 41712)\n",
      "0 columns renamed with suffix '_PRS313_Known'\n",
      "11 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 38762)\n",
      "9 columns renamed with suffix '_PRS313_Known'\n",
      "25 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 45799)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "18 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 36163)\n",
      "4 columns renamed with suffix '_PRS313_Known'\n",
      "10 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 33374)\n",
      "8 columns renamed with suffix '_PRS313_Known'\n",
      "13 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 27799)\n",
      "5 columns renamed with suffix '_PRS313_Known'\n",
      "9 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 33688)\n",
      "4 columns renamed with suffix '_PRS313_Known'\n",
      "14 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 32489)\n",
      "3 columns renamed with suffix '_PRS313_Known'\n",
      "16 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 30647)\n",
      "6 columns renamed with suffix '_PRS313_Known'\n",
      "11 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 22500)\n",
      "1 columns renamed with suffix '_PRS313_Known'\n",
      "4 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 20724)\n",
      "4 columns renamed with suffix '_PRS313_Known'\n",
      "4 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 19533)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "5 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 21222)\n",
      "3 columns renamed with suffix '_PRS313_Known'\n",
      "11 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 19395)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "7 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 18611)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "7 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 15455)\n",
      "0 columns renamed with suffix '_PRS313_Known'\n",
      "7 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 15458)\n",
      "1 columns renamed with suffix '_PRS313_Known'\n",
      "3 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 9040)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "2 columns added with suffix '_PRS313_Unknown'\n",
      "(2504, 9382)\n",
      "2 columns renamed with suffix '_PRS313_Known'\n",
      "9 columns added with suffix '_PRS313_Unknown'\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "ttandMePositions = \"../../../Data/23andMe_metadata_files/23andMeGenePanel.csv\"\n",
    "\n",
    "output_dir   = \"../../../Data/Raw_training_data_23andMe_union/\"\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "# Read the text file into a DataFrame\n",
    "ttAndMe = pd.read_csv(ttandMePositions, comment='#', header = 0)\n",
    "for chromosome_number in range(1, 23):\n",
    "    # Define directory path for the current chromosome\n",
    "    directory_path = f'../../../Data/23AndMePositionsUnion/chr{chromosome_number}/'\n",
    "    \n",
    "    # Process VCF files to get the DataFrame\n",
    "    df_23AndMe = process_vcf_files(directory_path)\n",
    "    \n",
    "    # Filter the PRS313 data to include only SNPs on the current chromosome\n",
    "    chromosome_N_snps = PRS313_df.loc[:, [col for col in PRS313_df.columns if col.startswith(f'chr{chromosome_number}_')]]\n",
    "    \n",
    "    # Identify columns present in both DataFrames\n",
    "    common_columns = df_23AndMe.columns.intersection(chromosome_N_snps.columns)\n",
    "    \n",
    "    # Rename the common columns in df_23AndMe\n",
    "    df_23AndMe.rename(columns={col: f'{col}_PRS313_Known' for col in common_columns}, inplace=True)\n",
    "    print(f\"{len(common_columns)} columns renamed with suffix '_PRS313_Known'\")\n",
    "    \n",
    "    # Identify columns in PRS313 not present in df_23AndMe using 'in_23andMe' row\n",
    "    in_23andMe = chromosome_N_snps.loc['in_23andMe'].astype(bool)\n",
    "    missing_columns = chromosome_N_snps.columns[~in_23andMe]\n",
    "    \n",
    "    # Add the missing columns to df_23AndMe with suffix '_PRS313_Unknown'\n",
    "    for column in missing_columns:\n",
    "        df_23AndMe[f'{column}_PRS313_Unknown'] = chromosome_N_snps[column]\n",
    "    print(f\"{len(missing_columns)} columns added with suffix '_PRS313_Unknown'\")\n",
    "    \n",
    "    # Save the resulting DataFrame to a parquet file\n",
    "    output_file = os.path.join(output_dir, f'23AndMe_PRS313_merged_chr{chromosome_number}.parquet')\n",
    "\n",
    "    df_23AndMe.to_parquet(output_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
