{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "def read_vcf_header(file_path):\n",
    "    \"\"\"Reads the VCF file header (column names) and returns it.\"\"\"\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('#CHROM'):\n",
    "                return line.strip().split('\\t')\n",
    "    return []\n",
    "\n",
    "def extract_snp_data(file_path, header):\n",
    "    \"\"\"Extracts SNP data from a VCF file.\"\"\"\n",
    "    snp_data = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith('#'):\n",
    "                fields = line.strip().split('\\t')\n",
    "                chrom, pos, ref, alt = fields[0], fields[1], fields[3], fields[4]\n",
    "                snp_id = f\"chr{chrom}_{pos}_{ref}_{alt}\"\n",
    "                genotype_data = fields[9:]  # Genotype data starts from the 10th column\n",
    "                snp_data.append((snp_id, genotype_data))\n",
    "    return snp_data\n",
    "\n",
    "def process_vcf_files(directory):\n",
    "    \"\"\"Processes all VCF files in the specified directory.\"\"\"\n",
    "    vcf_files = glob.glob(os.path.join(directory, '*.vcf'))\n",
    "    all_snp_data = []\n",
    "    patient_ids = None\n",
    "    \n",
    "    for file_path in vcf_files:\n",
    "        if patient_ids is None:\n",
    "            patient_ids = read_vcf_header(file_path)[9:]  # Assuming the first 9 columns are standard VCF columns\n",
    "        snp_data = extract_snp_data(file_path, patient_ids)\n",
    "\n",
    "        if (len(snp_data) == 0):\n",
    "            # print(\"No matches found at \", file_path)\n",
    "            continue\n",
    "        else:\n",
    "            for SNP in snp_data:\n",
    "                position_in_vcf = SNP[0].split(\"_\")[1]\n",
    "                file_path_position = file_path.split(\"_pos\")[1][:-4]\n",
    "                if (position_in_vcf == file_path_position):\n",
    "                    all_snp_data.extend(snp_data)\n",
    "                    # print(\"Adding\")\n",
    "                else:\n",
    "                    # Ignoring copy number variations\n",
    "                    # print(\"Ignoring\")\n",
    "                    continue\n",
    "\n",
    "    # Convert the collected SNP data to a DataFrame\n",
    "    data_dict = {snp_id: genotypes for snp_id, genotypes in all_snp_data}\n",
    "    df = pd.DataFrame(data_dict, index=patient_ids)\n",
    "    print(df.shape)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find intersection b/w 23AndMe and PRS313"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated PRS313_df:\n",
      "           chr2_192381934_C_T chr8_129199566_G_A chr8_29509616_A_C  \\\n",
      "HG00096                   1|1                0|1               1|1   \n",
      "HG00097                   1|1                1|0               1|1   \n",
      "HG00099                   1|1                0|0               0|0   \n",
      "HG00100                   0|1                0|0               0|1   \n",
      "HG00101                   1|1                0|0               1|1   \n",
      "...                       ...                ...               ...   \n",
      "NA21141                   1|1                0|0               0|0   \n",
      "NA21142                   1|0                1|0               0|1   \n",
      "NA21143                   1|1                0|0               1|1   \n",
      "NA21144                   1|1                0|0               0|1   \n",
      "in_23andMe               True              False              True   \n",
      "\n",
      "           chr2_121089731_T_C chr15_75750383_T_C chr18_24337424_C_G  \\\n",
      "HG00096                   1|0                1|0                1|1   \n",
      "HG00097                   0|0                0|0                0|1   \n",
      "HG00099                   0|1                0|0                1|1   \n",
      "HG00100                   0|0                0|0                0|1   \n",
      "HG00101                   1|0                1|1                1|1   \n",
      "...                       ...                ...                ...   \n",
      "NA21141                   0|0                0|1                1|1   \n",
      "NA21142                   0|0                1|0                0|1   \n",
      "NA21143                   0|0                1|1                0|1   \n",
      "NA21144                   0|0                0|1                1|0   \n",
      "in_23andMe               True              False               True   \n",
      "\n",
      "           chr5_131640536_A_G chr16_4008542_CAAAAA_C chr16_10706580_G_A  \\\n",
      "HG00096                   0|1                    1|1                0|0   \n",
      "HG00097                   1|1                    1|1                0|0   \n",
      "HG00099                   0|1                    1|1                0|0   \n",
      "HG00100                   1|0                    1|1                0|0   \n",
      "HG00101                   1|1                    1|1                0|0   \n",
      "...                       ...                    ...                ...   \n",
      "NA21141                   1|1                    1|1                0|0   \n",
      "NA21142                   0|1                    1|1                1|0   \n",
      "NA21143                   0|1                    1|1                0|0   \n",
      "NA21144                   0|0                    0|1                0|1   \n",
      "in_23andMe              False                  False              False   \n",
      "\n",
      "           chr9_110885479_C_T  ... chr12_120832146_C_T  \\\n",
      "HG00096                   1|1  ...                 0|0   \n",
      "HG00097                   1|0  ...                 1|0   \n",
      "HG00099                   1|0  ...                 0|0   \n",
      "HG00100                   1|1  ...                 0|1   \n",
      "HG00101                   0|1  ...                 0|1   \n",
      "...                       ...  ...                 ...   \n",
      "NA21141                   1|1  ...                 0|0   \n",
      "NA21142                   1|1  ...                 1|0   \n",
      "NA21143                   1|1  ...                 0|0   \n",
      "NA21144                   1|1  ...                 1|0   \n",
      "in_23andMe              False  ...                True   \n",
      "\n",
      "           chr3_63887449_T_TTG,TTGTG chr4_126752992_A_AAT,AATAT  \\\n",
      "HG00096                          0|0                        0|1   \n",
      "HG00097                          0|1                        1|1   \n",
      "HG00099                          0|0                        0|1   \n",
      "HG00100                          1|0                        1|0   \n",
      "HG00101                          0|1                        1|0   \n",
      "...                              ...                        ...   \n",
      "NA21141                          0|0                        1|1   \n",
      "NA21142                          0|1                        0|0   \n",
      "NA21143                          0|0                        1|0   \n",
      "NA21144                          0|0                        0|1   \n",
      "in_23andMe                     False                      False   \n",
      "\n",
      "           chr4_187503758_A_G,T chr4_84370124_TAA_TA,T chr5_176134882_T_A,C  \\\n",
      "HG00096                     0|1                    0|1                  0|1   \n",
      "HG00097                     0|1                    1|0                  1|0   \n",
      "HG00099                     1|1                    0|0                  0|0   \n",
      "HG00100                     1|0                    1|1                  1|1   \n",
      "HG00101                     1|0                    1|0                  1|0   \n",
      "...                         ...                    ...                  ...   \n",
      "NA21141                     0|0                    1|0                  1|1   \n",
      "NA21142                     0|1                    0|1                  1|1   \n",
      "NA21143                     0|1                    1|1                  0|0   \n",
      "NA21144                     0|0                    1|0                  1|1   \n",
      "in_23andMe                False                  False                False   \n",
      "\n",
      "           chr5_52679539_C_CA,CAA chr7_91459189_A_AT,ATT chr10_22861490_A_C,T  \\\n",
      "HG00096                       0|0                    1|0                  1|1   \n",
      "HG00097                       0|0                    1|0                  1|1   \n",
      "HG00099                       0|0                    1|1                  1|1   \n",
      "HG00100                       0|0                    0|1                  0|1   \n",
      "HG00101                       0|0                    0|0                  1|1   \n",
      "...                           ...                    ...                  ...   \n",
      "NA21141                       0|0                    1|1                  1|1   \n",
      "NA21142                       0|0                    0|1                  1|1   \n",
      "NA21143                       1|0                    1|0                  0|1   \n",
      "NA21144                       0|0                    1|0                  1|1   \n",
      "in_23andMe                  False                  False                False   \n",
      "\n",
      "           chr22_38583315_AAAAG_AAAAGAAAG,AAAAGAAAGAAAG,A  \n",
      "HG00096                                               0|0  \n",
      "HG00097                                               0|0  \n",
      "HG00099                                               1|1  \n",
      "HG00100                                               0|0  \n",
      "HG00101                                               0|0  \n",
      "...                                                   ...  \n",
      "NA21141                                               0|1  \n",
      "NA21142                                               0|1  \n",
      "NA21143                                               1|0  \n",
      "NA21144                                               1|0  \n",
      "in_23andMe                                          False  \n",
      "\n",
      "[2505 rows x 313 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define file paths\n",
    "ttAndMeFullPanel = \"../../../Data/23andMe_metadata_files/23andMeGenePanel.csv\"\n",
    "prs313_file = './concatenated_snps_processed.csv'\n",
    "\n",
    "# Read the CSV files into DataFrames\n",
    "ttAndMe = pd.read_csv(ttAndMeFullPanel)\n",
    "PRS313_df = pd.read_csv(prs313_file, index_col=0)\n",
    "\n",
    "# Extract chromosome and position information from PRS313_df columns\n",
    "PRS313_info = PRS313_df.columns.str.extract(r'chr(\\d+)_(\\d+)', expand=True)\n",
    "PRS313_info.columns = ['chromosome', 'position']\n",
    "PRS313_info['full_col'] = PRS313_df.columns\n",
    "\n",
    "# Convert chromosome and position columns to string type for comparison\n",
    "ttAndMe['chromosome'] = ttAndMe['chromosome'].astype(str)\n",
    "ttAndMe['position'] = ttAndMe['position'].astype(str)\n",
    "PRS313_info['chromosome'] = PRS313_info['chromosome'].astype(str)\n",
    "PRS313_info['position'] = PRS313_info['position'].astype(str)\n",
    "\n",
    "# Create a new row in PRS313_df to store the boolean values\n",
    "PRS313_df.loc['in_23andMe'] = False\n",
    "\n",
    "# Merge ttAndMe with PRS313_info to identify matching rows\n",
    "merged_df = pd.merge(ttAndMe, PRS313_info, how='inner', on=['chromosome', 'position'])\n",
    "\n",
    "# Set the 'in_23andMe' row to True for matching columns\n",
    "PRS313_df.loc['in_23andMe', merged_df['full_col']] = True\n",
    "\n",
    "# Print the updated PRS313_df DataFrame\n",
    "print(\"Updated PRS313_df:\")\n",
    "print(PRS313_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'>=' not supported between instances of 'str' and 'bool'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Find the max value in PRS313_df\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[43mPRS313_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/frame.py:11656\u001b[0m, in \u001b[0;36mDataFrame.max\u001b[0;34m(self, axis, skipna, numeric_only, **kwargs)\u001b[0m\n\u001b[1;32m  11648\u001b[0m \u001b[38;5;129m@doc\u001b[39m(make_doc(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m\"\u001b[39m, ndim\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m))\n\u001b[1;32m  11649\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmax\u001b[39m(\n\u001b[1;32m  11650\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  11654\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m  11655\u001b[0m ):\n\u001b[0;32m> 11656\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnumeric_only\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m  11657\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(result, Series):\n\u001b[1;32m  11658\u001b[0m         result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/generic.py:12404\u001b[0m, in \u001b[0;36mNDFrame.max\u001b[0;34m(self, axis, skipna, numeric_only, **kwargs)\u001b[0m\n\u001b[1;32m  12397\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmax\u001b[39m(\n\u001b[1;32m  12398\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m  12399\u001b[0m     axis: Axis \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  12402\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m  12403\u001b[0m ):\n\u001b[0;32m> 12404\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_stat_function\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m  12405\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmax\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12406\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnanops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnanmax\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12407\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12408\u001b[0m \u001b[43m        \u001b[49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12409\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnumeric_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12410\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  12411\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/generic.py:12377\u001b[0m, in \u001b[0;36mNDFrame._stat_function\u001b[0;34m(self, name, func, axis, skipna, numeric_only, **kwargs)\u001b[0m\n\u001b[1;32m  12373\u001b[0m nv\u001b[38;5;241m.\u001b[39mvalidate_func(name, (), kwargs)\n\u001b[1;32m  12375\u001b[0m validate_bool_kwarg(skipna, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mskipna\u001b[39m\u001b[38;5;124m\"\u001b[39m, none_allowed\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m> 12377\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reduce\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m  12378\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskipna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnumeric_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnumeric_only\u001b[49m\n\u001b[1;32m  12379\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/frame.py:11562\u001b[0m, in \u001b[0;36mDataFrame._reduce\u001b[0;34m(self, op, name, axis, skipna, numeric_only, filter_type, **kwds)\u001b[0m\n\u001b[1;32m  11558\u001b[0m     df \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mT\n\u001b[1;32m  11560\u001b[0m \u001b[38;5;66;03m# After possibly _get_data and transposing, we are now in the\u001b[39;00m\n\u001b[1;32m  11561\u001b[0m \u001b[38;5;66;03m#  simple case where we can use BlockManager.reduce\u001b[39;00m\n\u001b[0;32m> 11562\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreduce\u001b[49m\u001b[43m(\u001b[49m\u001b[43mblk_func\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m  11563\u001b[0m out \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39m_constructor_from_mgr(res, axes\u001b[38;5;241m=\u001b[39mres\u001b[38;5;241m.\u001b[39maxes)\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m  11564\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m out_dtype \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m out\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mboolean\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/internals/managers.py:1500\u001b[0m, in \u001b[0;36mBlockManager.reduce\u001b[0;34m(self, func)\u001b[0m\n\u001b[1;32m   1498\u001b[0m res_blocks: \u001b[38;5;28mlist\u001b[39m[Block] \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m   1499\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m blk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks:\n\u001b[0;32m-> 1500\u001b[0m     nbs \u001b[38;5;241m=\u001b[39m \u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreduce\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1501\u001b[0m     res_blocks\u001b[38;5;241m.\u001b[39mextend(nbs)\n\u001b[1;32m   1503\u001b[0m index \u001b[38;5;241m=\u001b[39m Index([\u001b[38;5;28;01mNone\u001b[39;00m])  \u001b[38;5;66;03m# placeholder\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/internals/blocks.py:404\u001b[0m, in \u001b[0;36mBlock.reduce\u001b[0;34m(self, func)\u001b[0m\n\u001b[1;32m    398\u001b[0m \u001b[38;5;129m@final\u001b[39m\n\u001b[1;32m    399\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mreduce\u001b[39m(\u001b[38;5;28mself\u001b[39m, func) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlist\u001b[39m[Block]:\n\u001b[1;32m    400\u001b[0m     \u001b[38;5;66;03m# We will apply the function and reshape the result into a single-row\u001b[39;00m\n\u001b[1;32m    401\u001b[0m     \u001b[38;5;66;03m#  Block with the same mgr_locs; squeezing will be done at a higher level\u001b[39;00m\n\u001b[1;32m    402\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m\n\u001b[0;32m--> 404\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    407\u001b[0m         res_values \u001b[38;5;241m=\u001b[39m result\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/frame.py:11481\u001b[0m, in \u001b[0;36mDataFrame._reduce.<locals>.blk_func\u001b[0;34m(values, axis)\u001b[0m\n\u001b[1;32m  11479\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39marray([result])\n\u001b[1;32m  11480\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m> 11481\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskipna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/nanops.py:147\u001b[0m, in \u001b[0;36mbottleneck_switch.__call__.<locals>.f\u001b[0;34m(values, axis, skipna, **kwds)\u001b[0m\n\u001b[1;32m    145\u001b[0m         result \u001b[38;5;241m=\u001b[39m alt(values, axis\u001b[38;5;241m=\u001b[39maxis, skipna\u001b[38;5;241m=\u001b[39mskipna, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    146\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 147\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43malt\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskipna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    149\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/nanops.py:404\u001b[0m, in \u001b[0;36m_datetimelike_compat.<locals>.new_func\u001b[0;34m(values, axis, skipna, mask, **kwargs)\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m datetimelike \u001b[38;5;129;01mand\u001b[39;00m mask \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    402\u001b[0m     mask \u001b[38;5;241m=\u001b[39m isna(values)\n\u001b[0;32m--> 404\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskipna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    406\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m datetimelike:\n\u001b[1;32m    407\u001b[0m     result \u001b[38;5;241m=\u001b[39m _wrap_results(result, orig_values\u001b[38;5;241m.\u001b[39mdtype, fill_value\u001b[38;5;241m=\u001b[39miNaT)\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/pandas/core/nanops.py:1098\u001b[0m, in \u001b[0;36m_nanminmax.<locals>.reduction\u001b[0;34m(values, axis, skipna, mask)\u001b[0m\n\u001b[1;32m   1093\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _na_for_min_count(values, axis)\n\u001b[1;32m   1095\u001b[0m values, mask \u001b[38;5;241m=\u001b[39m _get_values(\n\u001b[1;32m   1096\u001b[0m     values, skipna, fill_value_typ\u001b[38;5;241m=\u001b[39mfill_value_typ, mask\u001b[38;5;241m=\u001b[39mmask\n\u001b[1;32m   1097\u001b[0m )\n\u001b[0;32m-> 1098\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmeth\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1099\u001b[0m result \u001b[38;5;241m=\u001b[39m _maybe_null_out(result, axis, mask, values\u001b[38;5;241m.\u001b[39mshape)\n\u001b[1;32m   1100\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/Documents/medicaid/DeepImpute/.venv/lib/python3.11/site-packages/numpy/core/_methods.py:41\u001b[0m, in \u001b[0;36m_amax\u001b[0;34m(a, axis, out, keepdims, initial, where)\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_amax\u001b[39m(a, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, out\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, keepdims\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m     40\u001b[0m           initial\u001b[38;5;241m=\u001b[39m_NoValue, where\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[0;32m---> 41\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m umr_maximum(a, axis, \u001b[38;5;28;01mNone\u001b[39;00m, out, keepdims, initial, where)\n",
      "\u001b[0;31mTypeError\u001b[0m: '>=' not supported between instances of 'str' and 'bool'"
     ]
    }
   ],
   "source": [
    "# Find the max value in PRS313_df\n",
    "\n",
    "PRS313_df.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 53891)\n",
      "10 columns renamed with suffix '_PRS313_Known'\n",
      "20 columns added with suffix '_PRS313_Unknown'\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "ttandMePositions = \"../../../Data/23andMe_metadata_files/23andMeGenePanel.csv\"\n",
    "\n",
    "output_dir   = \"../../../Data/Raw_training_data_23andMe_union/\"\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "# Read the text file into a DataFrame\n",
    "ttAndMe = pd.read_csv(ttandMePositions, comment='#', header = 0)\n",
    "for chromosome_number in range(1, 23):\n",
    "    # Define directory path for the current chromosome\n",
    "    directory_path = f'../../../Data/23AndMePositionsUnion/chr{chromosome_number}/'\n",
    "    \n",
    "    # Process VCF files to get the DataFrame\n",
    "    df_23AndMe = process_vcf_files(directory_path)\n",
    "    \n",
    "    # Filter the PRS313 data to include only SNPs on the current chromosome\n",
    "    chromosome_N_snps = PRS313_df.loc[:, [col for col in PRS313_df.columns if col.startswith(f'chr{chromosome_number}_')]]\n",
    "    \n",
    "    # Identify columns present in both DataFrames\n",
    "    common_columns = df_23AndMe.columns.intersection(chromosome_N_snps.columns)\n",
    "    \n",
    "    # Rename the common columns in df_23AndMe\n",
    "    df_23AndMe.rename(columns={col: f'{col}_PRS313_Known' for col in common_columns}, inplace=True)\n",
    "    print(f\"{len(common_columns)} columns renamed with suffix '_PRS313_Known'\")\n",
    "    \n",
    "    # Identify columns in PRS313 not present in df_23AndMe using 'in_23andMe' row\n",
    "    in_23andMe = chromosome_N_snps.loc['in_23andMe'].astype(bool)\n",
    "    missing_columns = chromosome_N_snps.columns[~in_23andMe]\n",
    "    \n",
    "    # Add the missing columns to df_23AndMe with suffix '_PRS313_Unknown'\n",
    "    for column in missing_columns:\n",
    "        df_23AndMe[f'{column}_PRS313_Unknown'] = chromosome_N_snps[column]\n",
    "    print(f\"{len(missing_columns)} columns added with suffix '_PRS313_Unknown'\")\n",
    "    \n",
    "    # Save the resulting DataFrame to a parquet file\n",
    "    output_file = os.path.join(output_dir, f'23AndMe_PRS313_merged_chr{chromosome_number}.parquet')\n",
    "\n",
    "    df_23AndMe.to_parquet(output_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
