{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 1: R2 = 0.9018, IQS = 0.9160, Accuracy = 0.7928, AUC = 0.9540\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr1/individual_r2_scores_chr1.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr1/individual_iqs_scores_chr1.csv\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 2: R2 = 0.8922, IQS = 0.9075, Accuracy = 0.7762, AUC = 0.9537\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr2/individual_r2_scores_chr2.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr2/individual_iqs_scores_chr2.csv\n",
      "multiallelic SNP found at 3:63887449\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 3: R2 = 0.9464, IQS = 0.9593, Accuracy = 0.9232, AUC = 0.9730\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr3/individual_r2_scores_chr3.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr3/individual_iqs_scores_chr3.csv\n",
      "multiallelic SNP found at 4:84370124\n",
      "multiallelic SNP found at 4:126752992\n",
      "multiallelic SNP found at 4:187503758\n",
      "Chromosome 4: R2 = 0.1444, IQS = 0.8705, Accuracy = 0.8351, AUC = 0.9342\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr4/individual_r2_scores_chr4.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr4/individual_iqs_scores_chr4.csv\n",
      "multiallelic SNP found at 5:52679539\n",
      "multiallelic SNP found at 5:176134882\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 5: R2 = 0.8753, IQS = 0.9440, Accuracy = 0.8613, AUC = 0.9689\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr5/individual_r2_scores_chr5.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr5/individual_iqs_scores_chr5.csv\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 6: R2 = 0.9040, IQS = 0.9112, Accuracy = 0.7959, AUC = 0.9574\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr6/individual_r2_scores_chr6.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr6/individual_iqs_scores_chr6.csv\n",
      "multiallelic SNP found at 7:91459189\n",
      "Chromosome 7: R2 = -0.1620, IQS = 0.8816, Accuracy = 0.7358, AUC = 0.9382\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr7/individual_r2_scores_chr7.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr7/individual_iqs_scores_chr7.csv\n",
      "Chromosome 9: R2 = 0.9018, IQS = 0.9097, Accuracy = 0.7581, AUC = 0.9520\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr9/individual_r2_scores_chr9.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr9/individual_iqs_scores_chr9.csv\n",
      "multiallelic SNP found at 10:22861490\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 10: R2 = 0.8989, IQS = 0.9231, Accuracy = 0.8093, AUC = 0.9657\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr10/individual_r2_scores_chr10.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr10/individual_iqs_scores_chr10.csv\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 11: R2 = 0.9469, IQS = 0.9610, Accuracy = 0.9058, AUC = 0.9775\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr11/individual_r2_scores_chr11.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr11/individual_iqs_scores_chr11.csv\n",
      "Files for Chromosome 12 not found. Skipping...\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 13: R2 = 0.8174, IQS = 0.8768, Accuracy = 0.7948, AUC = 0.9388\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr13/individual_r2_scores_chr13.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr13/individual_iqs_scores_chr13.csv\n",
      "Files for Chromosome 14 not found. Skipping...\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 15: R2 = 0.8551, IQS = 0.8929, Accuracy = 0.8332, AUC = 0.9517\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr15/individual_r2_scores_chr15.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr15/individual_iqs_scores_chr15.csv\n",
      "Chromosome 16: R2 = 0.9097, IQS = 0.9191, Accuracy = 0.7618, AUC = 0.9601\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr16/individual_r2_scores_chr16.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr16/individual_iqs_scores_chr16.csv\n",
      "Files for Chromosome 17 not found. Skipping...\n",
      "Chromosome 18: R2 = 0.9695, IQS = 0.9677, Accuracy = 0.8428, AUC = 0.9839\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr18/individual_r2_scores_chr18.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr18/individual_iqs_scores_chr18.csv\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 19: R2 = 0.9625, IQS = 0.9671, Accuracy = 0.8942, AUC = 0.9861\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr19/individual_r2_scores_chr19.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr19/individual_iqs_scores_chr19.csv\n",
      "Chromosome 20: R2 = 0.8524, IQS = 0.8804, Accuracy = 0.7779, AUC = 0.9316\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr20/individual_r2_scores_chr20.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr20/individual_iqs_scores_chr20.csv\n",
      "Chromosome 21: R2 = 0.9555, IQS = 0.9530, Accuracy = 0.9356, AUC = 0.9797\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr21/individual_r2_scores_chr21.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr21/individual_iqs_scores_chr21.csv\n",
      "multiallelic SNP found at 22:38583315\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "All true labels are negative. Returning NaN.\n",
      "Chromosome 22: R2 = 0.9066, IQS = 0.9222, Accuracy = 0.8799, AUC = 0.9531\n",
      "Individual R^2 scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr22/individual_r2_scores_chr22.csv\n",
      "Individual IQS scores saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files//chr22/individual_iqs_scores_chr22.csv\n",
      "Overall R2: 0.8043\n",
      "Overall IQS: 0.9202\n",
      "Overall Accuracy: 0.8285\n",
      "Overall AUC: 0.9589\n",
      "Performance metrics saved at: ../../../Data/model_results_unphased_all_PRS/beagle/csv_files/performance_metrics.csv\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.metrics import roc_auc_score, r2_score, accuracy_score\n",
    "import csv\n",
    "\n",
    "def calculate_iqs_unphased(true_genotypes, imputed_dosages):\n",
    "    # Convert imputed dosages to discrete values\n",
    "    imputed_discrete = np.round(imputed_dosages).astype(int)\n",
    "\n",
    "    # Clip the imputed discrete values to be within the range of 0 to 2\n",
    "    imputed_discrete = np.clip(imputed_discrete, 0, 2)\n",
    "\n",
    "    # Create a contingency table\n",
    "    contingency_table = np.zeros((3, 3), dtype=int)\n",
    "\n",
    "    # Fill the contingency table\n",
    "    for true_geno, imputed_geno in zip(true_genotypes, imputed_discrete):\n",
    "        for true_allele, imputed_allele in zip(true_geno, imputed_geno):\n",
    "            contingency_table[int(true_allele), int(imputed_allele)] += 1\n",
    "\n",
    "    # Calculate the total number of genotypes\n",
    "    total_genotypes = np.sum(contingency_table)\n",
    "\n",
    "    # Calculate observed proportion of agreement (Po)\n",
    "    observed_agreement = np.trace(contingency_table) / total_genotypes\n",
    "\n",
    "    # Calculate marginal sums\n",
    "    row_marginals = np.sum(contingency_table, axis=1)\n",
    "    col_marginals = np.sum(contingency_table, axis=0)\n",
    "\n",
    "    # Calculate chance agreement (Pc)\n",
    "    chance_agreement = np.sum((row_marginals * col_marginals) / (total_genotypes ** 2))\n",
    "\n",
    "    # Calculate IQS\n",
    "    if chance_agreement == 1:  # To prevent division by zero in case of perfect chance agreement\n",
    "        iqs_score = 0\n",
    "    else:\n",
    "        iqs_score = (observed_agreement - chance_agreement) / (1 - chance_agreement)\n",
    "\n",
    "    return iqs_score\n",
    "\n",
    "def load_prs313_snps(prs313_path):\n",
    "    prs313_df = pd.read_excel(prs313_path)\n",
    "    prs313_snps = prs313_df[['Chromosome', 'Positionb', 'Reference Allele', 'Effect Allele']]\n",
    "    prs313_snps = prs313_snps.set_index(['Chromosome', 'Positionb'])\n",
    "    return prs313_snps\n",
    "\n",
    "def parse_vcf(vcf_path, prs313_snps, sample_ids):\n",
    "    allele_dosages = []\n",
    "    positions = []\n",
    "    with open(vcf_path, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                parts = line.strip().split(\"\\t\")\n",
    "                vcf_sample_ids = parts[9:]\n",
    "                if vcf_sample_ids != sample_ids:\n",
    "                    raise ValueError(\"Sample IDs in VCF file do not match the expected order\")\n",
    "            elif not line.startswith(\"#\"):\n",
    "                parts = line.strip().split(\"\\t\")\n",
    "                chrom = int(parts[0])\n",
    "                pos = int(parts[1])\n",
    "                ref = parts[3]\n",
    "                alt = parts[4]\n",
    "\n",
    "                if (chrom, pos) in prs313_snps.index:\n",
    "                    try:\n",
    "                        dosages = [float(gt.split(':')[1]) for gt in parts[9:]]\n",
    "                    except:\n",
    "                        vcf_alts = alt.split(\",\")\n",
    "                        PRS313_alt = prs313_snps.loc[(chrom, pos)][\"Effect Allele\"]\n",
    "                        print(f\"multiallelic SNP found at {chrom}:{pos}\")\n",
    "                        dosages = [float(gt.split(':')[1][vcf_alts.index(PRS313_alt) * 2]) for gt in parts[9:]]\n",
    "\n",
    "                    allele_dosages.append(dosages)\n",
    "                    positions.append(pos)\n",
    "    allele_dosages = np.array(allele_dosages).T  # Transpose to get samples as rows and SNPs as columns\n",
    "    return allele_dosages, positions\n",
    "\n",
    "def sort_true_labels(true_labels_df, positions):\n",
    "    sorted_columns = sorted(true_labels_df.columns, key=lambda x: int(x.split('_')[1]))\n",
    "    true_labels_df = true_labels_df[sorted_columns]\n",
    "    sorted_positions = sorted(positions)\n",
    "    try:\n",
    "        position_index = [sorted_positions.index(int(col.split('_')[1])) for col in sorted_columns]\n",
    "    except:\n",
    "        print(\"Error in label sorting\")\n",
    "    return true_labels_df.values[:, position_index]\n",
    "\n",
    "def safe_auc_metric(y_pred, y_true):\n",
    "    if y_true.sum() == 0:\n",
    "        print(\"All true labels are negative. Returning NaN.\")\n",
    "        return np.NaN\n",
    "    return roc_auc_score(y_true, y_pred)\n",
    "\n",
    "def evaluate_metrics(true_labels, imputed_dosages):\n",
    "    # Ensure both arrays have the same shape\n",
    "    if (imputed_dosages.shape != true_labels.shape):\n",
    "        raise ValueError(\"Shape of imputed dosages and true labels do not match\")\n",
    "    \n",
    "    # Calculate R2\n",
    "    r2_scores = r2_score(true_labels, imputed_dosages, multioutput='raw_values')\n",
    "    \n",
    "    # Calculate IQS\n",
    "    iqs_scores = np.array([calculate_iqs_unphased(true_labels[:, i].reshape(-1, 1), imputed_dosages[:, i].reshape(-1, 1)) for i in range(true_labels.shape[1])])\n",
    "    \n",
    "    # Calculate accuracy\n",
    "    correct_predictions = np.sum(true_labels == np.round(imputed_dosages))\n",
    "    total_predictions = true_labels.size\n",
    "    accuracy_scores = correct_predictions / total_predictions\n",
    "    \n",
    "    # Calculate AUC\n",
    "    individual_aucs = [\n",
    "        [\n",
    "            safe_auc_metric((imputed_dosages[:, j].round() == i).astype(float), (true_labels[:, j] == i).astype(float))\n",
    "            for i in range(3)\n",
    "        ]\n",
    "        for j in range(true_labels.shape[1])\n",
    "    ]\n",
    "    flattened_aucs = [auc for sublist in individual_aucs for auc in sublist]\n",
    "    auc_scores = np.nanmean(flattened_aucs)\n",
    "\n",
    "    # Flatten individual AUC scores into a single array for each SNP\n",
    "    individual_auc_scores = np.nanmean(np.array(individual_aucs), axis=1)\n",
    "\n",
    "    # Calculate individual accuracy for each SNP\n",
    "    individual_accuracy_scores = np.array([\n",
    "        ((true_labels[:, i] == np.round(imputed_dosages[:, i])).sum() / true_labels.shape[0])\n",
    "        for i in range(true_labels.shape[1])\n",
    "    ])\n",
    "\n",
    "    return r2_scores, iqs_scores, accuracy_scores, auc_scores, individual_auc_scores, individual_accuracy_scores\n",
    "\n",
    "def main():\n",
    "    true_labels_dir = \"../../../Data/y_test_labels_unphased/\"\n",
    "    imputed_vcf_dir = \"../../../Data/Beagle_imputed_output/\"\n",
    "    prs313_path = \"../../../Data/PRS313_with_23andMe.xlsx\"\n",
    "\n",
    "    # Load PRS313 SNPs\n",
    "    prs313_snps = load_prs313_snps(prs313_path)\n",
    "\n",
    "    # Initialize lists to store the performance metrics for each chromosome\n",
    "    all_r2_scores = []\n",
    "    all_iqs_scores = []\n",
    "    all_accuracy_scores = []\n",
    "    all_auc_scores = []\n",
    "    all_chromosomes = []\n",
    "\n",
    "    # Folders for saving results\n",
    "    output_folder = \"../../../Data/model_results_unphased_all_PRS/beagle/\"\n",
    "    csv_folder = output_folder + \"csv_files/\"\n",
    "\n",
    "    if not os.path.exists(csv_folder):\n",
    "        os.makedirs(csv_folder)\n",
    "\n",
    "    # Assuming chromosomes are numbered 1 to 22\n",
    "    for chr_num in range(1, 23):\n",
    "        sample_ids = None\n",
    "\n",
    "        true_labels_path = os.path.join(true_labels_dir, f\"chr{chr_num}_true_labels_y_test.csv\")\n",
    "        if os.path.exists(true_labels_path):\n",
    "            true_labels_df = pd.read_csv(true_labels_path, index_col=0)\n",
    "            sample_ids = true_labels_df.index.tolist()\n",
    "\n",
    "        if sample_ids is None:\n",
    "            raise ValueError(\"No true labels file found to extract sample IDs\")\n",
    "    \n",
    "        vcf_path = os.path.join(imputed_vcf_dir, f\"chr{chr_num}_specific_positions.vcf\")\n",
    "        \n",
    "        if os.path.exists(true_labels_path) and os.path.exists(vcf_path):\n",
    "            # Load imputed dosages from VCF\n",
    "            imputed_dosages, positions = parse_vcf(vcf_path, prs313_snps, sample_ids)\n",
    "\n",
    "            if (true_labels_df.shape[1] != len(positions)):\n",
    "                continue\n",
    "            else:\n",
    "                # Sort true labels to match imputed dosages order\n",
    "                true_labels = sort_true_labels(true_labels_df, positions)\n",
    "                r2_scores, iqs_scores, accuracy_scores, auc_scores, individual_auc_scores, individual_accuracy_scores = evaluate_metrics(true_labels, imputed_dosages)\n",
    "\n",
    "                # Evaluate metrics\n",
    "                all_r2_scores.append(r2_scores)\n",
    "                all_iqs_scores.append(iqs_scores)\n",
    "                all_accuracy_scores.append(accuracy_scores)\n",
    "                all_auc_scores.append(auc_scores)\n",
    "                all_chromosomes.append(chr_num)\n",
    "\n",
    "                print(f\"Chromosome {chr_num}: R2 = {np.mean(r2_scores):.4f}, IQS = {np.mean(iqs_scores):.4f}, Accuracy = {accuracy_scores:.4f}, AUC = {auc_scores:.4f}\")\n",
    "\n",
    "                # Get the names of the SNPs from the original dataframe\n",
    "                snp_names = true_labels_df.columns\n",
    "\n",
    "                individual_results_output_dir = f\"{csv_folder}/chr{chr_num}\"\n",
    "                os.makedirs(individual_results_output_dir, exist_ok=True)\n",
    "\n",
    "                # Save individual R^2 scores to a CSV file\n",
    "                csv_file = os.path.join(individual_results_output_dir, f'individual_r2_scores_chr{chr_num}.csv')\n",
    "\n",
    "                with open(csv_file, 'w', newline='') as file:\n",
    "                    writer = csv.writer(file)\n",
    "                    writer.writerow(['SNP', 'R2 Score'])\n",
    "                    for snp, r2_score in zip(snp_names, r2_scores):\n",
    "                        writer.writerow([snp, r2_score])\n",
    "\n",
    "                print(f\"Individual R^2 scores saved at: {csv_file}\")\n",
    "\n",
    "                # Save individual IQS scores to a CSV file\n",
    "                iqs_csv_file = os.path.join(individual_results_output_dir, f'individual_iqs_scores_chr{chr_num}.csv')\n",
    "\n",
    "                with open(iqs_csv_file, 'w', newline='') as file:\n",
    "                    writer = csv.writer(file)\n",
    "                    writer.writerow(['SNP', 'IQS Score'])\n",
    "                    for snp, iqs_score in zip(snp_names, iqs_scores):\n",
    "                        writer.writerow([snp, iqs_score])\n",
    "\n",
    "                print(f\"Individual IQS scores saved at: {iqs_csv_file}\")\n",
    "\n",
    "                # Save individual accuracy scores to a CSV file\n",
    "                accuracy_csv_file = os.path.join(individual_results_output_dir, f'individual_accuracy_scores_chr{chr_num}.csv')\n",
    "\n",
    "                with open(accuracy_csv_file, 'w', newline='') as file:\n",
    "                    writer = csv.writer(file)\n",
    "                    writer.writerow(['SNP', 'Accuracy Score'])\n",
    "                    for snp, accuracy_score in zip(snp_names, individual_accuracy_scores):\n",
    "                        writer.writerow([snp, accuracy_score])\n",
    "\n",
    "                print(f\"Individual Accuracy scores saved at: {accuracy_csv_file}\")\n",
    "\n",
    "                # Save individual AUC scores to a CSV file\n",
    "                auc_csv_file = os.path.join(individual_results_output_dir, f'individual_auc_scores_chr{chr_num}.csv')\n",
    "\n",
    "                with open(auc_csv_file, 'w', newline='') as file:\n",
    "                    writer = csv.writer(file)\n",
    "                    writer.writerow(['SNP', 'AUC Score'])\n",
    "                    for snp, auc_score in zip(snp_names, individual_auc_scores):\n",
    "                        writer.writerow([snp, auc_score])\n",
    "\n",
    "                print(f\"Individual AUC scores saved at: {auc_csv_file}\")\n",
    "\n",
    "        else:\n",
    "            print(f\"Files for Chromosome {chr_num} not found. Skipping...\")\n",
    "\n",
    "    # Drop nan\n",
    "    no_nan_accuracies = [x for x in all_accuracy_scores if x != 0]\n",
    "    overall_accuracy = np.mean(no_nan_accuracies)\n",
    "\n",
    "    # Calculate overall metrics\n",
    "    overall_r2 = [np.mean(scores) for scores in all_r2_scores]\n",
    "    overall_iqs = [np.mean(scores) for scores in all_iqs_scores]\n",
    "    overall_auc = np.mean(all_auc_scores)\n",
    "\n",
    "    print(f\"Overall R2: {np.mean(overall_r2):.4f}\")\n",
    "    print(f\"Overall IQS: {np.mean(overall_iqs):.4f}\")\n",
    "    print(f\"Overall Accuracy: {overall_accuracy:.4f}\")\n",
    "    print(f\"Overall AUC: {overall_auc:.4f}\")\n",
    "\n",
    "    # Save the overall performance metrics to a CSV file\n",
    "    performance_df = pd.DataFrame({\n",
    "        \"Chromosome\": all_chromosomes,\n",
    "        'R2 Score': overall_r2,\n",
    "        'IQS Score': overall_iqs,\n",
    "        'Accuracy Score': all_accuracy_scores,\n",
    "        'AUC Score': all_auc_scores\n",
    "    })\n",
    "\n",
    "    performance_csv_file = os.path.join(csv_folder, 'performance_metrics.csv')\n",
    "    performance_df.to_csv(performance_csv_file, index=False)\n",
    "    print(f\"Performance metrics saved at: {performance_csv_file}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
